---
title: "MSK SPECTRUM data freeze: major cell type compositions and embeddings"
author: "Florian Uhlitz"
date: "`r Sys.Date()`"
output: 
  html_document: 
    df_print: kable
    number_sections: yes
    toc: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}

knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = F)

```

```{r chunk_010}

library(magrittr)
library(tidyverse)
library(Seurat)
library(readxl)
library(cowplot)
library(colorblindr)
library(viridis)
library(magick, lib.loc = "/home/uhlitzf/miniconda3/lib/R/library")
library(ggpubr)

theme_cowplot2 <- function(...) {
  theme_cowplot(font_size = 16, font_family = "sans", ...) %+replace%
    theme(strip.background = element_blank(),
          panel.background = element_rect(fill = "transparent", color = NA),
          plot.background = element_rect(fill = "transparent", color = NA),
          panel.border = element_blank())
}
theme_set(theme_cowplot2())

```

# Overview

```{r chunk_030}

## set colors -------------------------------------------------

clrs <- yaml::read_yaml("/home/uhlitzf/spectrum_explore/_data/small/signatures/hgsc_v5_colors.yaml") %>% 
  lapply(function(x) purrr::map_depth(x, purrr::vec_depth(x)-2, unlist))

names(clrs$patient_id) <- str_remove_all(names(clrs$patient_id), "SPECTRUM-OV-")
names(clrs$cell_type) <- str_replace_all(names(clrs$cell_type), "\\.", " ")

cell_type_super_lookup <- c(
  B.cell = "Immune",
  Plasma.cell = "Immune",
  T.cell = "Immune", 
  Monocyte = "Immune", 
  Myeloid.cell = "Immune", 
  Dendritic.cell = "Immune", 
  Endothelial.cell = "Stromal",
  Fibroblast = "Stromal", 
  Ovarian.cancer.cell = "Stromal", 
  Other = "Other"
)

super_sets <- c("Ovarian.cancer.cell", "Myeloid.super", "T.cell", "B.super", "Fibroblast", "Endothelial.cell")

## load data --------------------------------------

## load super set embeddings
seu_tbl_super <- lapply(super_sets, function(x) read_tsv(paste0("/work/shah/uhlitzf/data/SPECTRUM/freeze/v5/", x, "_embedding.tsv"))) %>% bind_rows

## load cohort embeddings
seu_tbl <- read_tsv("/work/shah/isabl_data_lake/analyses/16/52/1652/cells.tsv") %>% 
  mutate(cell_type = ifelse(cell_type == "Monocyte", "Myeloid.cell", cell_type))

## load cell phase scores
# cell_phase_tbl <- lapply(super_sets, function(x) read_rds(paste0("/work/shah/isabl_data_lake/analyses/16/52/1652/celltypes/", x, "_processed.rds"))) %>% 
#   lapply(function(x) bind_cols(cell_id = colnames(x), FetchData(x, c("Phase", "S.Score", "G2M.Score")))) %>% 
#   bind_rows
# write_tsv(cell_phase_tbl, "/work/shah/uhlitzf/data/SPECTRUM/freeze/v5/cell_phase.tsv")
cell_phase_tbl <- read_tsv("/work/shah/uhlitzf/data/SPECTRUM/freeze/v5/cell_phase.tsv")

## load meta data
meta_tbl <- read_excel("/home/uhlitzf/spectrum_explore/_data/small/MSK SPECTRUM - Single cell RNA-seq.xlsx", sheet = 3)

## load wgs meta data
format_sig_tbl <- . %>% 
  select(patient_id, consensus_signature, BRCA1_impact, BRCA2_impact) %>% 
  bind_rows(tibble(patient_id = unique(sort(meta_tbl$patient_id[!(meta_tbl$patient_id %in% .$patient_id)])), consensus_signature = "NA")) %>% 
  mutate(consensus_signature = ordered(consensus_signature, levels = names(clrs$consensus_signature))) %>% 
  arrange(consensus_signature) %>% 
  distinct(patient_id, .keep_all = T)

signature_tbl <- read_tsv("/home/uhlitzf/spectrum_explore/_data/small/mutational_signatures_summary.tsv") %>% 
  format_sig_tbl

signature_tbl_old <- read_tsv("/home/uhlitzf/spectrum_explore/_data/small/mutational_signatures_summary_old.tsv") %>% 
  format_sig_tbl

signature_tbl <- signature_tbl %>% 
  select(patient_id, consensus_signature) %>% 
  left_join(select(signature_tbl_old, -consensus_signature), by = "patient_id") %>% 
  mutate(BRCA1_status = ifelse(is.na(BRCA1_impact), NA, "MUT"),
         BRCA2_status = ifelse(is.na(BRCA2_impact), NA, "MUT"),
         BRCA_status = ifelse(!is.na(BRCA1_status), "BRCA1", ifelse(!is.na(BRCA2_status), "BRCA2", NA)))

## load consensus data
round3 <- function(x) round(x, 3)
subtype_names <- c(IMR_consensus = "Immunoreactive", DIF_consensus = "Differentiated", PRO_consensus = "Proliferative", MES_consensus = "Mesenchymal")
consOV_tbl <- list.files("/work/shah/uhlitzf/data/SPECTRUM/freeze/v5/consensusOV/", full.names = T, pattern = "freeze-consensusOV") %>% 
  lapply(read_tsv) %>% 
  bind_rows() %>% 
  mutate_if(.predicate = is.numeric, .funs = round3) %>% 
  mutate(consensusOV_label = subtype_names[consensusOV]) %>% 
  mutate(consensusOV = ordered(consensusOV_label, levels = names(clrs$consensusOV))) %>% 
  select(-consensusOV_label)

# ## load infer cnv info
# infercnv_files <- paste0(list.files("/work/shah/shih/infercnv", pattern = "external", full.names = T), "/infercnv.meta.csv")
# 
# coltypes <- cols(
#   barcode = col_character(),
#   cp_score = col_double(),
#   cp_length = col_double(),
#   clone = col_character(),
#   cell_type = col_character()
# )
# 
# cnv_tbl <- lapply(infercnv_files, read_csv, col_types = coltypes) %>% 
#   bind_rows %>% 
#   dplyr::rename(cell_type_cnv = cell_type,
#                 cell_id = barcode) %>% 
#   filter(clone != "reference") %>% 
#   mutate(pid_clone = paste0(str_sub(cell_id, 13, 15), "_", clone)) %>% 
#   group_by(pid_clone, cell_type_cnv) %>% 
#   mutate(n_cell_type = length(cell_type_cnv)) %>% 
#   group_by(pid_clone) %>% 
#   mutate(cp_score_mean = mean(cp_score), 
#          cp_length_mean = mean(cp_length),
#          clone_cell_type = cell_type_cnv[which.max(n_cell_type)]
#          ) %>% 
#   mutate(cnv_gate = cp_length_mean > 37.5) %>% 
#   select(-n_cell_type) %>% 
#   mutate(pid_cancer_clone = ifelse(cnv_gate, pid_clone, NA)) %>% 
#   ungroup

# cnv_tbl %>%
#   distinct(pid_clone, cp_score_mean, cp_length_mean, clone_cell_type) %>%
#   separate(pid_clone, into = c("pid", "clone")) %>% 
#   ggplot() +
#   geom_point(aes(cp_score_mean, log2(cp_length_mean), color = clone_cell_type)) +
#   geom_hline(yintercept = log2(37.5))


## join data -----------------------------------

u1_tbl <- select(seu_tbl, cell_id, contains("_1")) %>% 
  gather(reduction, UMAP_1, -cell_id) %>% 
  mutate(reduction = str_remove_all(reduction, "_1"))
u2_tbl <- select(seu_tbl, cell_id, contains("_2")) %>% 
  gather(reduction, UMAP_2, -cell_id) %>% 
  mutate(reduction = str_remove_all(reduction, "_2"))

emb_tbl <- left_join(u1_tbl, u2_tbl, by = c("cell_id", "reduction"))

seu_tbl_full <- emb_tbl %>% 
  left_join(select(seu_tbl, cell_id, nCount_RNA, nFeature_RNA, percent.mt,
                   sample, cell_type, doublet, doublet_score), by = "cell_id") %>% 
  left_join(select(meta_tbl, sample = isabl_id, patient_id, contains("tumor"), 
                   sort_parameters, therapy, qc_status), by = "sample") %>% 
  #filter(qc_status != "Fail") %>% 
  mutate(tumor_supersite = str_replace_all(tumor_supersite, "Upper Quadrant", "UQ")) %>% 
  mutate(cell_type_super = cell_type_super_lookup[cell_type]) %>% 
  mutate(cell_type = ordered(str_replace_all(cell_type, "\\.", " "), 
                             levels = names(clrs$cell_type)),
         patient_id_short = str_remove_all(patient_id, "SPECTRUM-OV-"),
         tumor_supersite = ordered(tumor_supersite, 
                                   levels = names(clrs$tumor_supersite)),
         sort_short = str_remove_all(sort_parameters, "singlet, live, "), 
         pid_rx = paste0(patient_id_short, ", ", therapy)) %>% 
  mutate(sort_short_x = ifelse(sort_short == "U" & cell_type_super == "Immune", 
                               "CD45+", ifelse(sort_short == "U" & cell_type_super == "Stromal", "CD45-", sort_short))) %>% 
  #left_join(cnv_tbl, by = "cell_id") %>% 
  left_join(signature_tbl, by = "patient_id") %>% 
  arrange(tumor_supersite) %>% 
  mutate(tumor_subsite = ordered(tumor_subsite, levels = unique(tumor_subsite))) %>% 
  left_join(consOV_tbl, by = "cell_id") %>% 
  left_join(select(seu_tbl_super, cluster_label, cell_id), by = "cell_id") %>% 
  left_join(cell_phase_tbl, by = "cell_id")

patient_id_short <- sort(unique(seu_tbl_full$patient_id_short))

seu_tbl_full50 <- seu_tbl_full %>% 
  # sample_n(10000) %>% 
  filter(reduction == "umap50", cell_type != "Other", therapy == "pre-Rx", !str_detect(sample, "UNKNOWN"))

```

# Fig1 / S1

## UMAPs

```{r chunk_040, fig.width=7.25, fig.height=8}

dpi <- 72

base_umap <- ggplot(seu_tbl_full50) +
  coord_fixed() +
  NoAxes() +
  theme(legend.position = c(0, 1),
        legend.justification = c("left", "top"),
        legend.box.just = "left",
        legend.margin = margin(1, 1, 1, 1),
        #panel.border = element_rect(linetype = 1, color = "black", size = 1),
        legend.text = element_text(size = 14, margin = margin(0, 10, 0, 0)),
        legend.spacing.x = unit(0, "npc"), 
        legend.title = element_blank(),
        plot.title = element_text(hjust = 0.5, vjust = 0.5, face = "plain", size = 22))

pt.size <- 0.1
pt.size.mini <- 0.01
pt.alpha <- 0.05
pt.alpha.mini <- 0.01

median_tbl <- seu_tbl_full50 %>% 
  filter(cell_type != "Other") %>% 
  group_by(cell_type, reduction) %>% 
  summarise(UMAP_1 = median(UMAP_1),
            UMAP_2 = median(UMAP_2))

umap_site <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = tumor_supersite), size = pt.size.mini, alpha = pt.alpha.mini, raster.dpi = 300) +
  scale_color_manual(values = clrs$tumor_supersite) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1),
                              ncol = 1, 
                              label.position = "right"))

umap_cell_type <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = fct_rev(cell_type)), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  scale_color_manual(values = clrs$cell_type) +
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1),
                              ncol = 1, 
                              label.position = "right")) +
  theme(legend.text = element_text(size = 14, margin = margin(0, 10, 9.5, 0), color = "white"),
        legend.spacing.x = unit(0, "npc"))

umap_cell_type_mini <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = cell_type), size = pt.size.mini, alpha = pt.alpha.mini, raster.dpi = dpi) +
  scale_color_manual(values = clrs$cell_type) +
  guides(color = F)

umap_cell_type_legend <- cowplot::get_legend(umap_cell_type)

umap_patient <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = patient_id_short), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  scale_color_manual(values = clrs$patient_id) + 
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1),
                              ncol = 3, 
                              label.position = "right")) +
  theme(legend.text = element_text(size = 14, margin = margin(0, 10, 0, 0)),
        legend.spacing.x = unit(0, "npc"))

umap_patient_mini <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = patient_id_short), size = pt.size.mini, alpha = pt.alpha.mini, raster.dpi = dpi) +
  scale_color_manual(values = clrs$patient_id) +
  guides(color = F)

umap_patient_legend <- cowplot::get_legend(umap_patient)

umap_consOV <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = consensusOV), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  scale_color_manual(values = clrs$consensusOV) + 
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1),
                              ncol = 1, 
                              label.position = "right")) +
  theme(legend.text = element_text(size = 14, margin = margin(0, 10, 0, 0)),
        legend.spacing.x = unit(0, "npc"))

umap_consOV_mini <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = consensusOV), size = pt.size.mini, alpha = pt.alpha.mini, raster.dpi = dpi) +
  scale_color_manual(values = clrs$consensusOV) +
  guides(color = F)

umap_consOV_legend <- cowplot::get_legend(umap_consOV)

umap_mutsig <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = consensus_signature), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  scale_color_manual(values = clrs$consensus_signature) + 
  guides(color = guide_legend(override.aes = list(size = 4, alpha = 1),
                              ncol = 1, 
                              label.position = "right")) +
  theme(legend.text = element_text(size = 14, margin = margin(0, 10, 0, 0)),
        legend.spacing.x = unit(0, "npc"))

umap_mutsig_mini <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = consensus_signature), size = pt.size.mini, alpha = pt.alpha, raster.dpi = dpi) +
  scale_color_manual(values = clrs$consensus_signature) +
  guides(color = F)

umap_mutsig_legend <- cowplot::get_legend(umap_mutsig)

continuous_umap_layers <- list(
  scale_color_viridis_c(),
  guides(color = guide_colorbar(label.position = "right",
                                title.position = "top",
                                title.hjust = 0,
                                title.vjust = 1,
                                direction = "vertical")),
  theme(legend.title = element_text(face = "bold", size = 14),
        legend.key.height = unit(0.05, "npc"),
        legend.key.width = unit(0.03, "npc"),
        legend.position = c(-0.07, 1),
        legend.justification = c("left", "top"),
        plot.title = element_text(hjust = 0.5, vjust = 0.5, face = "plain", size = 18)))

umap_umis <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = log2(nCount_RNA)), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  labs(title = "log2(#UMIs)", color = "") +
  continuous_umap_layers

umap_genes <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = log2(nFeature_RNA)), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  labs(title = "log2(#Genes)", color = "") +
  continuous_umap_layers

umap_mitos <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = percent.mt), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  labs(title = "MT reads [%]", color = "") +
  continuous_umap_layers

umap_doublet_score <- base_umap + 
  ggrastr::geom_point_rast(aes(UMAP_1, UMAP_2, color = doublet_score), size = pt.size, alpha = pt.alpha, raster.dpi = dpi) +
  labs(title = "Doublet score", color = "") +
  continuous_umap_layers

```

## total numbers 

```{r chunk_050, fig.width=2.75, fig.height=8}

total_numbers_cell_type <- seu_tbl_full50 %>%
  filter(cell_type != "Other", therapy == "pre-Rx") %>%
  group_by(cell_type) %>%
  tally %>%
  arrange(n) %>%
  mutate(cell_type_label = paste0(cell_type, " (", format(n, trim=T, big.mark=","), ")")) %>% 
  #mutate(cell_type = ordered(cell_type, levels = rev(as.character(cell_type)))) %>%
  ggplot() +
  geom_bar(aes(cell_type, n, fill = cell_type), stat = "identity", width = 0.3) +
  geom_text(aes(cell_type, 1, label = cell_type_label), hjust = 0, vjust = 0, nudge_x = 0.2) +
  scale_fill_manual(values = clrs$cell_type) +
  guides(fill = F) +
  coord_flip() +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 200000),
                     #labels = c("0", "100k", "200k"),
                     labels = function(x) parse(text = paste0(x/(10^5), "10^5")),
                     breaks = c(0, 100000, 200000)) +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(face = "plain")) +
  labs(y = "Number of cells")

# total_numbers_cell_type

# total_numbers_site <- seu_tbl_full50 %>%
#   filter(cell_type != "Other", therapy == "pre-Rx") %>%
#   distinct(tumor_supersite, sample) %>%
#   group_by(tumor_supersite) %>%
#   tally %>%
#   arrange(n) %>%
#   mutate(tumor_supersite = ordered(tumor_supersite, levels = rev(as.character(tumor_supersite)))) %>%
#   ggplot(aes(tumor_supersite, n, fill = tumor_supersite)) +
#   geom_bar(stat = "identity") +
#   scale_fill_manual(values = clrs$tumor_supersite) +
#   guides(fill = F) +
#   scale_y_continuous(expand = c(0, 0), limits = c(0, 75), breaks = c(0, 25, 50, 75)) +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
#         axis.title.x = element_blank(),
#         plot.title = element_text(face = "plain")) +
#   labs(y = "Number of samples",
#        title = "Site")
# 
# total_numbers_grid <- plot_grid(total_numbers_cell_type, total_numbers_site, ncol = 1, align = "v")
# 
# total_numbers_grid
# ggsave("/home/uhlitzf/spectrumanalysis/results/scrna/cell_type_compositions/fig/001_total_numbers.pdf", total_numbers_grid, width = 3.3, height = 5.5)

```

```{r chunk_060, fig.width=15, fig.height=9}

arrow <- arrow(angle = 20, type = "closed", length = unit(0.1, "npc"))
umap_coord_anno <- ggplot(tibble(group = c("UMAP1", "UMAP2"),
                                 x = c(0, 0), xend = c(1, 0),
                                 y = c(0, 0), yend = c(0, 1),
                                 lx = c(0.5, -0.15), ly = c(-0.15, 0.5),
                                 angle = c(0, 90))) +
  geom_segment(aes(x, y, xend = xend, yend = yend, group = group),
               arrow = arrow, size = 1, lineend = "round") +
  geom_text(aes(lx, ly, label = group, angle = angle), size = 4) +
  theme_void() +
  coord_fixed(xlim = c(-0.3, 1), ylim = c(-0.3, 1))

add_umap_coord <- function(gg_obj) {
  p <- ggdraw() + 
    draw_plot(gg_obj, x = 0, y = 0, width = 1, height = 1) +
    draw_plot(umap_coord_anno, x = -0.015, y = -0.02, width = 0.2, height = 0.2)
  return(p)
}

xf <- 1.25

ggdraw() +
  draw_plot(add_umap_coord(umap_site), x = 0, y = 0, width = 0.66/xf, height = 1) +
  draw_plot(umap_cell_type_mini, x = 0.6/xf, y = 0.66, width = 0.33/xf, height = 0.33) +
  draw_plot(umap_patient_mini, x = 0.6/xf, y = 0.33, width = 0.33/xf, height = 0.33) +
  draw_plot(umap_consOV_mini, x = 0.6/xf, y = 0, width = 0.33/xf, height = 0.33) +
  draw_grob(umap_cell_type_legend, x = 0.72, y = 0, hjust = 0, vjust = 0) +
  draw_plot(total_numbers_cell_type, x = 0.725, y = 0.64, width = 0.18, height = 0.35) +
  draw_grob(umap_patient_legend, x = 0.72, y = -0.35, hjust = 0, vjust = 0) +
  draw_grob(umap_consOV_legend, x = 0.72, y = -0.68, hjust = 0, vjust = 0) +
  draw_label("Site", x = 0.35/xf, y = 0.98, size = 20) +
  draw_label("Cell type", x = 0.65/xf, y = 0.98, size = 20, hjust = 0) +
  draw_label("Patient", x = 0.65/xf, y = 0.65, size = 20, hjust = 0) +
  draw_label("TCGA", x = 0.65/xf, y = 0.31, size = 20, hjust = 0)

ggsave(filename = "_fig/001x_umaps.png", width = 15, height = 9)
ggsave(filename = "_fig/001x_umaps.pdf", width = 15, height = 9)

```

```{r chunk_061, fig.width=18, fig.height=7}

xf <- 1.33

ggdraw() +
  draw_plot(add_umap_coord(umap_site), x = 0, y = 0, width = 0.5/xf, height = 1) +
  draw_plot(umap_patient_mini, x = 0.5/xf, y = 0.5, width = 0.25/xf, height = 0.5) +
  draw_plot(umap_cell_type_mini, x = 0.5/xf, y = 0, width = 0.25/xf, height = 0.5) +
  draw_plot(umap_mutsig_mini, x = 0.94/xf, y = 0.5, width = 0.25/xf, height = 0.5) +
  draw_plot(umap_consOV_mini, x = 0.94/xf, y = 0, width = 0.25/xf, height = 0.5) +
  draw_grob(umap_cell_type_legend, x = 0.56, y = -0.53, hjust = 0, vjust = 0) +
  draw_plot(total_numbers_cell_type, x = 0.565, y = 0.005, width = 0.13, height = 0.45) +
  draw_grob(umap_patient_legend, x = 0.56, y = -0.05, hjust = 0, vjust = 0) +
  draw_grob(umap_mutsig_legend, x = 0.9, y = -0.05, hjust = 0, vjust = 0) +
  draw_grob(umap_consOV_legend, x = 0.9, y = -0.55, hjust = 0, vjust = 0) +
  draw_label("Site", x = 0.26/xf, y = 0.98, size = 20) +
  draw_label("Patient", x = 0.52/xf, y = 0.98, size = 20, hjust = 0) +
  draw_label("Cell type", x = 0.52/xf, y = 0.49, size = 20, hjust = 0) +
  draw_label("Signature", x = 0.95/xf, y = 0.98, size = 20, hjust = 0) +
  draw_label("TCGA", x = 0.95/xf, y = 0.49, size = 20, hjust = 0)

ggsave(filename = "_fig/001x_umaps_mutsig.png", width = 16, height = 7)
ggsave(filename = "_fig/001x_umaps_mutsig.pdf", width = 18, height = 7)


```


## cell type compositions per site


```{r chunk_080, fig.width=8, fig.height=4}

comp_tbl_sample_sort <- seu_tbl_full50 %>% 
  filter(therapy == "pre-Rx", cell_type != "Other") %>% 
  group_by(tumor_subsite, tumor_supersite, patient_id_short, 
           therapy, sort_short_x, consensus_signature, 
           BRCA1_status, BRCA2_status, BRCA_status, cell_type) %>% 
  tally %>% 
  group_by(tumor_subsite, tumor_supersite, patient_id_short, 
           therapy, sort_short_x, consensus_signature,
           BRCA1_status, BRCA2_status, BRCA_status) %>% 
  mutate(nrel = n/sum(n)*100,
         log10n = log10(n))

expand_samples <- function(pid, x, tbl) {
  filter(tbl, patient_id_short == pid) %>% 
    {expand_grid(unique(.$tumor_subsite), unique(.[[x]]), unique(.$therapy))} %>%
    setNames(c("tumor_subsite", x, "therapy")) %>% 
    mutate(patient_id_short = pid, n = NA, nrel = NA, log10n = NA)
}

samples_expanded <- bind_rows(lapply(unique(comp_tbl_sample_sort$patient_id_short), expand_samples, "sort_short_x", comp_tbl_sample_sort))

comp_tbl_sample_sort <- bind_rows(comp_tbl_sample_sort, samples_expanded) %>% 
  ungroup %>% 
  mutate(label_supersite = "Site",
         label_therapy = "Rx",
         label_mutsig = "Signature") %>% 
  mutate(tumor_supersite = ordered(tumor_supersite, levels = rev(names(clrs$tumor_supersite)))) %>% 
  mutate(sample_id = paste(tumor_subsite, patient_id_short, therapy, sort_short_x)) %>% 
  mutate(patient_id_short = ordered(patient_id_short, levels = str_sub(signature_tbl$patient_id, 13, 15))) %>% 
  arrange(cell_type != "Myeloid cell", -nrel) %>% 
  mutate(sample_id_Myeloid.cell = ordered(sample_id, levels = unique(sample_id))) %>% 
  arrange(cell_type != "T cell", nrel) %>% 
  mutate(sample_id_T.cell = ordered(sample_id, levels = unique(sample_id))) %>% 
  group_by(cell_type, sort_short_x) %>% 
  mutate(rank_T.cell = row_number(sample_id_T.cell)) %>% 
  mutate(rank_T.cell = scales::rescale(rank_T.cell, c(-1, 1))) %>% 
  ungroup %>% 
  arrange(cell_type != "B cell", -nrel) %>% 
  mutate(sample_id_B.cell = ordered(sample_id, levels = unique(sample_id))) %>% 
  arrange(cell_type != "Plasma cell", -nrel) %>% 
  mutate(sample_id_Plasma.cell = ordered(sample_id, levels = unique(sample_id))) %>% 
  arrange(cell_type != "Fibroblast", -nrel) %>% 
  mutate(sample_id_Fibroblast = ordered(sample_id, levels = unique(sample_id))) %>% 
  arrange(cell_type != "Ovarian cancer cell", -nrel) %>% 
  mutate(sample_id_Ovarian.cancer.cell = ordered(sample_id, levels = unique(sample_id))) %>% 
  group_by(cell_type, sort_short_x) %>% 
  mutate(rank_Ovarian.cancer.cell = row_number(sample_id_Ovarian.cancer.cell)) %>% 
  mutate(rank_Ovarian.cancer.cell = scales::rescale(rank_Ovarian.cancer.cell, c(-1, 1))) %>% 
  ungroup

comp_tbl_consOV <- seu_tbl_full50 %>% 
  filter(therapy == "pre-Rx", cell_type != "Other") %>% 
  group_by(tumor_subsite, tumor_supersite, patient_id_short, 
           therapy, sort_short_x, consensus_signature, consensusOV) %>% 
  tally %>% 
  group_by(tumor_subsite, tumor_supersite, patient_id_short, 
           therapy, sort_short_x, consensus_signature) %>% 
  mutate(nrel = n/sum(n)*100,
         log10n = log10(n))

samples_expanded_consOV <- bind_rows(lapply(unique(comp_tbl_consOV$patient_id_short), expand_samples, "sort_short_x", comp_tbl_consOV))

comp_tbl_consOV <- bind_rows(comp_tbl_consOV, samples_expanded_consOV) %>% 
  ungroup %>% 
  mutate(label_supersite = "Site",
         label_therapy = "Rx",
         label_mutsig = "Signature") %>% 
  mutate(tumor_supersite = ordered(tumor_supersite, levels = rev(names(clrs$tumor_supersite)))) %>% 
  mutate(sample_id = paste(tumor_subsite, patient_id_short, therapy, sort_short_x)) %>% 
  mutate(patient_id_short = ordered(patient_id_short, levels = str_sub(signature_tbl$patient_id, 13, 15))) %>% 
  arrange(consensusOV != "Immunoreactive", nrel) %>% 
  mutate(sample_id_IR = ordered(sample_id, levels = unique(sample_id))) %>% 
  group_by(consensusOV, sort_short_x) %>% 
  mutate(rank_IR = row_number(sample_id_IR)) %>% 
  mutate(rank_IR = scales::rescale(rank_IR, c(-1, 1))) %>% 
  ungroup %>% 
  arrange(consensusOV != "Mesenchymal", -nrel) %>% 
  mutate(sample_id_MC = ordered(sample_id, levels = unique(sample_id))) %>% 
  group_by(consensusOV, sort_short_x) %>% 
  mutate(rank_MC = row_number(sample_id_MC)) %>% 
  mutate(rank_MC = scales::rescale(rank_MC, c(-1, 1))) %>% 
  ungroup %>% 
  left_join(comp_tbl_sample_sort %>% filter(cell_type == "T cell") %>% distinct(sample_id, sample_id_T.cell, rank_T.cell), by = "sample_id") %>% 
  left_join(comp_tbl_sample_sort %>% filter(cell_type == "Ovarian cancer cell") %>% distinct(sample_id, sample_id_Ovarian.cancer.cell, rank_Ovarian.cancer.cell), by = "sample_id")

```

```{r chunk_085, fig.width=3, fig.height=3}

cd45n_ranks <- comp_tbl_sample_sort %>% 
  filter(cell_type == "Ovarian cancer cell", sort_short_x == "CD45-") %>% 
  select(tumor_supersite, consensus_signature, rank = rank_Ovarian.cancer.cell)

cd45p_ranks <- comp_tbl_sample_sort %>% 
  filter(cell_type == "T cell", sort_short_x == "CD45+") %>% 
  select(tumor_supersite, consensus_signature, rank = rank_T.cell)

wilcoxon_wrapper <- function(rank_tbl, test_var = "tumor_supersite", 
                             test_value, filter_ascites = F) {
  if (filter_ascites) rank_tbl <- filter(rank_tbl, tumor_supersite != "Ascites")
  result <- tryCatch({
    x <- filter(rank_tbl)$rank
    y <- filter(rank_tbl, rank_tbl[[test_var]] == test_value)$rank
    wilcox.test(x = x, y = y, paired = F)$p.value
  }, error = function(e) NA)
  return(result)
}

site_names <- unique(comp_tbl_sample_sort$tumor_supersite)
mut_names <- unique(comp_tbl_sample_sort$consensus_signature)

cd45n_tests_site <- site_names %>% 
  lapply(function(x) wilcoxon_wrapper(cd45n_ranks, test_value = x)) %>% 
  setNames(site_names) %>% 
  enframe("comparison", "pval") %>% 
  unnest %>% 
  na.omit

cd45p_tests_site <- site_names %>% 
  lapply(function(x) wilcoxon_wrapper(cd45p_ranks, test_value = x)) %>% 
  setNames(site_names) %>% 
  enframe("comparison", "pval") %>% 
  unnest %>% 
  na.omit

cd45n_tests_mut <- mut_names %>% 
  lapply(function(x) wilcoxon_wrapper(cd45n_ranks, test_var = "consensus_signature",
                                      test_value = x, filter_ascites = T)) %>% 
  setNames(mut_names) %>% 
  enframe("comparison", "pval") %>% 
  unnest %>% 
  na.omit

cd45p_tests_mut <- mut_names %>% 
  lapply(function(x) wilcoxon_wrapper(cd45p_ranks, test_var = "consensus_signature",
                                      test_value = x, filter_ascites = T)) %>% 
  setNames(mut_names) %>% 
  enframe("comparison", "pval") %>% 
  unnest %>% 
  na.omit

rank_tests <- bind_rows(list(`CD45+` = cd45p_tests_site, `CD45-` = cd45n_tests_site), .id = "sort_short_x") %>% 
  mutate(fdr = p.adjust(pval, method = "BH"))

```

```{r chunk_090, fig.width=12, fig.height=8}

comp_plot_wrapper <- function(comp_tbl = comp_tbl_sample_sort, cts, y = "nrel", x = "tumor_subsite", fill = "cell_type", switch = NULL, facet = F, exclude_ascites = F) {
  if (facet == F) comp_tbl <- filter(comp_tbl, !is.na(tumor_supersite))
  if (exclude_ascites == T) comp_tbl <- filter(comp_tbl, tumor_supersite != "Ascites")
  if (y == "nrel") ylab <- paste0("% cells")
  if (y == "log10n") ylab <- paste0("# cells")
  if (y == "n") ylab <- paste0("# cells")
  p <- ggplot(filter(comp_tbl, sort_short_x == cts), 
              aes_string(x, y, fill = fill)) + 
    scale_fill_manual(values = clrs[[fill]]) + 
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          axis.title.x = element_blank(),
          axis.line.x = element_blank(),
          axis.title.y = element_text(margin = margin(0, -1, 0, 1, unit = "npc")),
          strip.text.y = element_blank(),
          strip.background.y = element_blank(),
          plot.margin = grid::unit(c(0.04, 0.01, 0.02, 0.01), "npc")) + 
    #scale_x_discrete(expand = c(0, 0)) + 
    labs(x = "",
         y = ylab) +
    guides(fill = FALSE)
  if (facet == "tumor_supersite") p <- p + 
    facet_grid(sort_short_x~tumor_supersite, 
               space = "free", scales = "free", switch = switch)
  if (facet == "patient_id_short") p <- p + 
    facet_grid(sort_short_x~patient_id_short, 
               space = "free", scales = "free", switch = switch)
  if (facet == F) p <- p + 
    facet_grid(sort_short_x~., 
               space = "free", scales = "free", switch = switch)
  if (y == "nrel") p <- p + 
    geom_bar(stat = "identity", width = 1) +
    scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 50, 100), 
                       labels = c("0", "50", "100")) + 
    theme(strip.text.x = element_blank(),
          strip.background.x = element_blank())
  if (y == "log10n") p <- p + 
    geom_bar(stat = "identity", position = position_dodge2(preserve = "single"), width = 1) +
    scale_y_continuous(expand = c(0, 0), 
                       breaks = c(1, 2, 3),
                       limits = c(0, 4),
                       labels = function(x) parse(text = paste("10^", x))) +
    expand_limits(y = c(0, 4)) +
    theme(panel.grid.major.y = element_line(linetype = 1, color = "grey90", size = 0.5))
  if (y == "n") p <- p + 
    geom_bar(stat = "identity", position = position_stack(), width = 1) +
    scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 2500, 5000, 7500),
                       limits = c(0, 7500),
                       labels = c("", "2500", "5000", "7500")) +
    expand_limits(y = c(0, 7500)) +
    theme(panel.grid.major.y = element_line(linetype = 1, color = "grey90", size = 0.5),
          plot.title = element_text(face = "plain", size = 14, 
                                    hjust = 0.5, vjust = 0.5)) +
    labs(title = cts)
  return(p)
} 

common_label_layers <- list(
  theme(axis.title.y = element_blank(),
        strip.text = element_blank(),
        strip.background = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks = element_blank(), 
        axis.line.x = element_blank(),
        plot.margin = ggplot2::margin(0.02, 0.01, 0.02, 0.01, "npc")), 
  scale_y_discrete(expand = c(0, 0)),
  guides(fill = F, color = F)
)

comp_label <- function(x = "tumor_subsite", y = "label_supersite", fill = "tumor_supersite", facet = F, cts = c("CD45-", "CD45+")) {
  comp_tbl <- comp_tbl_sample_sort
  if (facet == F) comp_tbl <- filter(comp_tbl, !is.na(tumor_supersite))
  p <- ggplot(distinct(filter(comp_tbl, therapy == "pre-Rx", sort_short_x %in% cts), tumor_subsite, tumor_supersite, consensus_signature, label_supersite, label_mutsig, patient_id_short, sample_id_Myeloid.cell, sample_id_T.cell, rank_T.cell, sample_id_Fibroblast, sample_id_Ovarian.cancer.cell, rank_Ovarian.cancer.cell, sample_id_B.cell, sample_id_Plasma.cell), 
              aes_string(x, y, fill = fill)) + 
    geom_tile() + 
    scale_fill_manual(values = clrs[[fill]]) + 
    common_label_layers
  if (facet == "patient_id_short") p <- p + facet_grid(label_supersite~patient_id_short, 
                                      space = "free", scales = "free")
  if (facet == "tumor_supersite") p <- p + facet_grid(label_supersite~tumor_supersite, 
                                      space = "free", scales = "free")
  if (facet == F) p <- p + facet_grid(label_supersite~., 
                                      space = "free", scales = "free")
  return(p)
}

comp_label_boxplot <- function(comp_tbl = comp_tbl_sample_sort, x = "tumor_supersite", y = "rank_T.cell", facet = T, cts = c("CD45-", "CD45+"), clr = "tumor_supersite", ct = names(clrs$cell_type), pvals = F, exclude_ascites = F) {
  if (facet == F) comp_tbl <- filter(comp_tbl, !is.na(tumor_supersite))
  if (exclude_ascites == T) comp_tbl <- filter(comp_tbl, tumor_supersite != "Ascites")
  rank_tests_sub <- filter(rank_tests, sort_short_x %in% cts, pval < 0.05) %>% 
    mutate(pstar = "*")
  required_columns <- c("tumor_subsite", "tumor_supersite", "consensus_signature", "BRCA1_status", "BRCA2_status", "BRCA_status", "label_supersite", "patient_id_short", "sample_id_Myeloid.cell", "sample_id_T.cell", "rank_T.cell", "sample_id_Fibroblast", "sample_id_Ovarian.cancer.cell", "rank_Ovarian.cancer.cell", "sample_id_B.cell", "sample_id_Plasma.cell", "rank_IR", "sample_id_IR", "rank_MC", "sample_id_MC", "sample_cd45p")
  missing_columns <- required_columns[!(required_columns %in% colnames(comp_tbl))]
  comp_tbl <- bind_cols(comp_tbl, matrix(rep("distinct_helper", n = length(missing_columns)), ncol = length(missing_columns)) %>% set_colnames(missing_columns) %>% as_tibble())
  p <- ggplot(
    distinct(filter(comp_tbl, therapy == "pre-Rx", sort_short_x %in% cts, cell_type %in% ct), tumor_subsite, tumor_supersite, consensus_signature, BRCA1_status, BRCA2_status, BRCA_status, label_supersite, patient_id_short, sample_id_Myeloid.cell, sample_id_T.cell, rank_T.cell, sample_id_Fibroblast, sample_id_Ovarian.cancer.cell, rank_Ovarian.cancer.cell, sample_id_B.cell, sample_id_Plasma.cell, rank_IR, sample_id_IR, rank_MC, sample_id_MC, sample_cd45p)
     # distinct(filter(comp_tbl, therapy == "pre-Rx", sort_short_x %in% cts, cell_type %in% ct), -n, -log10n, -nrel)
    ) + 
    geom_tile(aes_string(x, y, fill = clr), 
              alpha = 0.8) +
    geom_boxplot(aes_string(x, y, fill = clr),
                 width = 0.35, size = 2.5, color = "white", outlier.shape = NA) +
    geom_boxplot(aes_string(x, y, color = clr),
                 width = 0.35, size = 1, fill = "white", outlier.shape = NA) +
    geom_hline(yintercept = 0, linetype = 2, size = 0.5, color = "black") +
    scale_fill_manual(values = clrs[[clr]]) + 
    scale_color_manual(values = clrs[[clr]]) + 
    coord_flip(clip = "off", ylim = c(-1.01, 1.01)) +
    theme(axis.title.y = element_blank(),
          strip.text = element_blank(),
          strip.background = element_blank(),
          plot.margin = ggplot2::margin(0.04, 0.05, 0.02, 0.01, "npc")) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0), breaks = c(-1, 0, 1)) +
    guides(fill = F, color = F) +
    labs(y = "Scaled rank")
  if (facet == T) p <- p + facet_grid(label_supersite~patient_id_short, 
                                      space = "free", scales = "free")
  if (facet == F) p <- p + facet_grid(label_supersite~., 
                                      space = "free", scales = "free")
  if (pvals == T) p <- p + 
    geom_text(aes(x = comparison, y = 1.05, label = pstar), data = rank_tests_sub, hjust = 0.5, vjust = 1, size = 5, angle = 90) #+
    # geom_segment(aes(x = "Adnexa", xend = comparison, y = 1.07, yend = 1.07), data = rank_tests_sub)
  return(p)
}


# ggdraw() + draw_plot(comp_label_boxplot(x = "tumor_supersite", y = "rank_T.cell", facet = F, cts = "CD45+", ct = "T cell", pvals = T), width = 0.9)

hist_plot_wrapper <- function(cts, x = "nrel") {
  if (x == "nrel") {
    xlab <- paste0("% cells")
    bw <- 5
  }
  if (x == "log10n") {
    xlab <- paste0("# cells")
    bw <- 0.2
  }
  p <- ggplot(filter(comp_tbl_sample_sort, sort_short_x == cts, !is.na(cell_type))) +
    ggridges::geom_density_ridges(
      aes_string(x, "cell_type", fill = "cell_type"), color = "black",
      stat = "binline", binwidth = bw, scale = 3) +
    facet_grid(label_supersite~label_therapy, 
               space = "free", scales = "free") +
    scale_fill_manual(values = clrs$cell_type) + 
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          axis.title.y = element_blank(),
          axis.line.y = element_blank(),
          strip.text = element_blank(),
          strip.background = element_blank(),
          plot.margin = ggplot2::margin(0.02, 0, 0.02, 0, "npc")) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_discrete(expand = c(0, 0)) +
    guides(fill = FALSE) +
    labs(x = xlab)
  if (x == "log10n") p <- p + expand_limits(x = c(0, 4)) + 
    scale_x_continuous(expand = c(0, 0), 
                       labels = function(x) parse(text = paste("10^", x)))
  return(p)
}

```

## vector rank plot

```{r chunk_095, fig.width=12, fig.height=8}

prim_ranks_wrapper <- function(x, prim_meta) {
  x %>% 
    distinct(patient_id_short, tumor_supersite, tumor_subsite, rank, sort_short_x) %>% 
    group_by(patient_id_short, sort_short_x) %>% 
    mutate(median_rank = mean(rank)) %>% 
    mutate(prim_meta = prim_meta) %>% 
    ungroup %>% 
    select(patient_id_short, sort_short_x, rank, median_rank, prim_meta, tumor_supersite) %>% 
    ungroup
} 

prim_ranks_cd45p <- filter(comp_tbl_sample_sort, sort_short_x == "CD45+") %>% 
  filter((tumor_supersite %in% c("Adnexa")), cell_type == "T cell") %>% 
  mutate(rank = rank_T.cell) %>% 
  prim_ranks_wrapper("Adnexa")

prim_ranks_cd45n <- filter(comp_tbl_sample_sort, sort_short_x == "CD45-") %>% 
  filter((tumor_supersite %in% c("Adnexa")), cell_type == "Ovarian cancer cell") %>% 
  mutate(rank = rank_Ovarian.cancer.cell) %>% 
  prim_ranks_wrapper("Adnexa")

meta_ranks_cd45p <- filter(comp_tbl_sample_sort, sort_short_x == "CD45+") %>% 
  filter(!(tumor_supersite %in% c("Adnexa")), cell_type == "T cell") %>% 
  mutate(rank = rank_T.cell) %>% 
  prim_ranks_wrapper("Non-adnexa")

meta_ranks_cd45n <- filter(comp_tbl_sample_sort, sort_short_x == "CD45-") %>% 
  filter(!(tumor_supersite %in% c("Adnexa")), cell_type == "Ovarian cancer cell") %>% 
  mutate(rank = rank_Ovarian.cancer.cell) %>% 
  prim_ranks_wrapper("Non-adnexa")


# asci_ranks_cd45p <- filter(comp_tbl_sample_sort, sort_short_x == "CD45+") %>% 
#   filter((tumor_supersite %in% c("Ascites")), cell_type == "T cell") %>% 
#   mutate(rank = rank_T.cell) %>% 
#   prim_ranks_wrapper("Ascites")
# 
# asci_ranks_cd45n <- filter(comp_tbl_sample_sort, sort_short_x == "CD45-") %>% 
#   filter((tumor_supersite %in% c("Ascites")), cell_type == "Ovarian cancer cell") %>% 
#   mutate(rank = rank_Ovarian.cancer.cell) %>% 
#   prim_ranks_wrapper("Ascites")

median_rank_wrapper <- . %>%
  select(-rank, -tumor_supersite) %>% 
  spread(prim_meta, median_rank) %>%
  ungroup() %>% 
  mutate(direction = case_when(
    sort_short_x == "CD45+" & Adnexa > `Non-adnexa` ~ "Depleted",
    sort_short_x == "CD45+" & Adnexa < `Non-adnexa` ~ "Enriched",
    sort_short_x == "CD45-" & Adnexa > `Non-adnexa` ~ "Depleted",
    sort_short_x == "CD45-" & Adnexa < `Non-adnexa` ~ "Enriched",
    sort_short_x == "mpIF" & Adnexa > `Non-adnexa` ~ "Depleted",
    sort_short_x == "mpIF" & Adnexa < `Non-adnexa` ~ "Enriched",
    sort_short_x == "Tumor" & Adnexa > `Non-adnexa` ~ "Depleted",
    sort_short_x == "Tumor" & Adnexa < `Non-adnexa` ~ "Enriched",
    sort_short_x == "Stroma" & Adnexa > `Non-adnexa` ~ "Depleted",
    sort_short_x == "Stroma" & Adnexa < `Non-adnexa` ~ "Enriched"
  )) %>% 
  arrange(
    #desc(direction), 
    Adnexa) %>% 
  na.omit() %>% 
  mutate(patient_id_short = ordered(patient_id_short, levels = unique(patient_id_short)),
         shape_helper = "Adnexa")

median_rank_tbl_cd45p <- bind_rows(distinct(prim_ranks_cd45p, median_rank, .keep_all = T), 
                                   distinct(meta_ranks_cd45p, median_rank, .keep_all = T)) %>% 
  median_rank_wrapper


median_rank_tbl_cd45n <- bind_rows(distinct(prim_ranks_cd45n, median_rank, .keep_all = T), 
                                   distinct(meta_ranks_cd45n, median_rank, .keep_all = T)) %>% 
  median_rank_wrapper


rank_tbl_cd45p <- bind_rows(prim_ranks_cd45p, meta_ranks_cd45p) %>% 
  mutate(prim_meta = ifelse(tumor_supersite == "Ascites", "Ascites", prim_meta),
         patient_id_short = ordered(patient_id_short, levels = levels(median_rank_tbl_cd45p$patient_id_short)))

rank_tbl_cd45n <- bind_rows(prim_ranks_cd45n, meta_ranks_cd45n) %>% 
  mutate(prim_meta = ifelse(tumor_supersite == "Ascites", "Ascites", prim_meta),
         patient_id_short = ordered(patient_id_short, levels = levels(median_rank_tbl_cd45n$patient_id_short)))


vector_patient_lvls_cd45p <- levels(rank_tbl_cd45p$patient_id_short)
vector_patient_lvls_cd45p[as.logical(1:length(vector_patient_lvls_cd45p) %% 2)] <- paste0(vector_patient_lvls_cd45p[as.logical(1:length(vector_patient_lvls_cd45p) %% 2)], "       ")

vector_patient_lvls_cd45n <- levels(rank_tbl_cd45n$patient_id_short)
vector_patient_lvls_cd45n[as.logical(1:length(vector_patient_lvls_cd45n) %% 2)] <- paste0(vector_patient_lvls_cd45n[as.logical(1:length(vector_patient_lvls_cd45n) %% 2)], "       ")

vector_rank_list_cd45p <- list(rank_tbl = rank_tbl_cd45p, 
                               median_rank_tbl = median_rank_tbl_cd45p, 
                               vector_patient_lvls = vector_patient_lvls_cd45p)

vector_rank_list_cd45n <- list(rank_tbl = rank_tbl_cd45n, 
                               median_rank_tbl = median_rank_tbl_cd45n, 
                               vector_patient_lvls = vector_patient_lvls_cd45n)

vector_patient_plot_wrapper <- function(data_list, vector_clrs) {
  
  ggplot() + 
    geom_vline(xintercept = 0, linetype = 2) +
    geom_point(aes(rank, patient_id_short, shape = prim_meta), color = "black",
               data = na.omit(data_list$rank_tbl)) +
    geom_segment(aes(x = Adnexa, xend = `Non-adnexa`, 
                     y = patient_id_short, yend = patient_id_short, 
                     color = direction), 
                 arrow = arrow(type = "open", length = unit(0.04, "npc")),
                 data = data_list$median_rank_tbl) +
    scale_shape_manual(values = c(`Adnexa` = 19, `Non-adnexa` = 21, `Ascites` = 4)) + 
    scale_color_manual(values = vector_clrs) +
    scale_x_continuous(expand = c(0, 0), breaks = c(-1, 0, 1), labels = c(-1, 0, 1)) +
    # scale_y_discrete(labels = data_list$vector_patient_lvls) +
    labs(x = "Scaled rank", y = "Patient", shape = "Sample", color = "Fraction\nin non-adnexa") + 
    theme(axis.title.y = element_blank(),
          strip.text = element_blank(),
          plot.margin = ggplot2::margin(0.02, 0.01, 0.02, 0.01, "npc")) +
    facet_grid(sort_short_x~., 
               space = "free", scales = "free") +
    expand_limits(x = c(-1, 1))
  
} 

vector_plot_cd45p <- vector_patient_plot_wrapper(vector_rank_list_cd45p, c("#ff7f00", "#33a02c"))
vector_plot_cd45n <- vector_patient_plot_wrapper(vector_rank_list_cd45n, c("#e31a1c", "#6a3d9a"))

```


```{r chunk_100, fig.width=3, fig.height=6}

remove_xaxis <- theme(axis.title.x = element_blank(),
                      axis.text.x = element_blank(),
                      axis.ticks.x = element_blank(),
                      axis.line.x = element_blank())

remove_guides <- guides(color = F, fill = F, shape = F, alpha = F)

pcomp1 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "n", x = "sample_id_T.cell", facet = F)
pcomp2 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "nrel", x = "sample_id_T.cell", facet = F)

pcomp_grid_p <- plot_grid(pcomp1, pcomp2,
                          comp_label_boxplot(x = "tumor_supersite", y = "rank_T.cell", facet = F, cts = "CD45+", ct = "T cell", pvals = T) + remove_xaxis, 
                          vector_plot_cd45p + remove_guides, 
                          ncol = 1, align = "v", 
                          rel_heights = c(0.15, 0.15, 0.25, 0.45))

ggsave("_fig/001x_cd45p_sorted_comp.pdf", pcomp_grid_p, width = 3, height = 6)

pcomp1 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "n", x = "sample_id_Ovarian.cancer.cell", facet = F)
pcomp2 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "nrel", x = "sample_id_Ovarian.cancer.cell", facet = F)

pcomp_grid_n <- plot_grid(pcomp1, pcomp2,
                          comp_label_boxplot(x = "tumor_supersite", y = "rank_Ovarian.cancer.cell", facet = F, cts = "CD45-", ct = "Ovarian cancer cell") + remove_xaxis,
                          vector_plot_cd45n + remove_guides, 
                          ncol = 1, align = "v", 
                          rel_heights = c(0.15, 0.15, 0.25, 0.45))

ggsave("_fig/001x_cd45n_sorted_comp.pdf", pcomp_grid_n, width = 3, height = 6)

```

```{r chunk_101, fig.width=3, fig.height=4}

pcomp1 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "n", x = "sample_id_T.cell", facet = F, exclude_ascites = T)
pcomp2 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "nrel", x = "sample_id_T.cell", facet = F, exclude_ascites = T)

pcomp_grid_p_mut <- plot_grid(pcomp1, pcomp2,
                          comp_label_boxplot(x = "consensus_signature", clr = "consensus_signature", y = "rank_T.cell", facet = F, cts = "CD45+", ct = "T cell", pvals = F, exclude_ascites = T), 
                          ncol = 1, align = "v", 
                          rel_heights = c(0.25, 0.25, 0.5))

ggsave("_fig/001x_cd45p_sorted_comp_mut.pdf", pcomp_grid_p_mut, width = 3, height = 4)

pcomp1 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "n", x = "sample_id_Ovarian.cancer.cell", facet = F, exclude_ascites = T)
pcomp2 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "nrel", x = "sample_id_Ovarian.cancer.cell", facet = F, exclude_ascites = T)

pcomp_grid_n_mut <- plot_grid(pcomp1, pcomp2,
                        comp_label_boxplot(x = "consensus_signature", clr = "consensus_signature", y = "rank_Ovarian.cancer.cell", facet = F, cts = "CD45-", ct = "Ovarian cancer cell", exclude_ascites = T), 
                        ncol = 1, align = "v", 
                        rel_heights = c(0.25, 0.25, 0.5))

ggsave("_fig/001x_cd45n_sorted_comp_mut.pdf", pcomp_grid_n_mut, width = 3, height = 4)


```

```{r chunk_102, fig.width=3, fig.height=5}

pcomp1 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "n", x = "sample_id_T.cell", facet = F)
pcomp2 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "nrel", x = "sample_id_T.cell", facet = F)

pcomp_grid_p_full <- plot_grid(pcomp1, pcomp2,
                          comp_label_boxplot(x = "tumor_supersite", y = "rank_T.cell", facet = F, cts = "CD45+", ct = "T cell", pvals = T) + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank()), 
                          comp_label_boxplot(x = "consensus_signature", clr = "consensus_signature", y = "rank_T.cell", facet = F, cts = "CD45+", ct = "T cell", pvals = F, exclude_ascites = T),
                          ncol = 1, align = "v", 
                          rel_heights = c(0.17, 0.17, 0.28, 0.28))

ggsave("_fig/001x_cd45p_sorted_comp_full.pdf", pcomp_grid_p_full, width = 3, height = 5)

pcomp1 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "n", x = "sample_id_Ovarian.cancer.cell", facet = F)
pcomp2 <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "nrel", x = "sample_id_Ovarian.cancer.cell", facet = F)

pcomp_grid_n_full <- plot_grid(pcomp1, pcomp2,
                          comp_label_boxplot(x = "tumor_supersite", y = "rank_Ovarian.cancer.cell", facet = F, cts = "CD45-", ct = "Ovarian cancer cell") + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank()), 
                          comp_label_boxplot(x = "consensus_signature", clr = "consensus_signature", y = "rank_Ovarian.cancer.cell", facet = F, cts = "CD45-", ct = "Ovarian cancer cell", exclude_ascites = T),
                          ncol = 1, align = "v", 
                          rel_heights = c(0.17, 0.17, 0.28, 0.28))

ggsave("_fig/001x_cd45n_sorted_comp_full.pdf", pcomp_grid_n_full, width = 3, height = 5)


```

```{r chunk_103, fig.width=3, fig.height=4}

## per patient composition plots

pcomp1p_cell_n <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "n", x = "sample_id_T.cell", facet = "patient_id_short") + theme(strip.text = element_text(angle = 90))
pcomp2p_cell_nrel <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45+", "nrel", x = "sample_id_T.cell", facet = "patient_id_short") + labs(y = "% cells\n(cell type)")
pcomp3p_consOV <- comp_plot_wrapper(comp_tbl_consOV, "CD45+", "nrel", fill = "consensusOV", x = "sample_id_T.cell", facet = "patient_id_short", exclude_ascites = F) + labs(y = "% cells\n(TCGA)")

pcompxp_label_site <- comp_label(x = "sample_id_T.cell", y = "label_supersite", fill = "tumor_supersite", cts = "CD45+", facet = "patient_id_short")
pcompxp_label_mutsig <- comp_label(x = "sample_id_T.cell", y = "label_mutsig", fill = "consensus_signature", cts = "CD45+", facet = "patient_id_short")

pcomp1n_cell_n <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "n", x = "sample_id_Ovarian.cancer.cell", facet = "patient_id_short") + theme(strip.text = element_text(angle = 90))
pcomp2n_cell_nrel <- comp_plot_wrapper(comp_tbl_sample_sort, "CD45-", "nrel", x = "sample_id_Ovarian.cancer.cell", facet = "patient_id_short") + labs(y = "% cells\n(cell type)")
pcomp3n_consOV <- comp_plot_wrapper(comp_tbl_consOV, "CD45-", "nrel", fill = "consensusOV", x = "sample_id_Ovarian.cancer.cell", facet = "patient_id_short", exclude_ascites = F) + labs(y = "% cells\n(TCGA)")

pcompxn_label_site <- comp_label(x = "sample_id_Ovarian.cancer.cell", y = "label_supersite", fill = "tumor_supersite", cts = "CD45-", facet = "patient_id_short")
pcompxn_label_mutsig <- comp_label(x = "sample_id_Ovarian.cancer.cell", y = "label_mutsig", fill = "consensus_signature", cts = "CD45-", facet = "patient_id_short")

comp_grid_per_patient_p <- plot_grid(
  pcomp1p_cell_n, 
  pcomp2p_cell_nrel, 
  pcomp3p_consOV, 
  pcompxp_label_site, 
  pcompxp_label_mutsig, 
  ncol = 1, align = "v", axis = "x", rel_heights = c(0.3, 0.24, 0.24, 0.06, 0.06)
)

comp_grid_per_patient_n <- plot_grid(
  pcomp1n_cell_n, 
  pcomp2n_cell_nrel, 
  pcomp3n_consOV, 
  pcompxn_label_site, 
  pcompxn_label_mutsig, 
  ncol = 1, align = "v", axis = "x", rel_heights = c(0.3, 0.24, 0.24, 0.06, 0.06)
)
  
```

```{r chunk_104, fig.width=3, fig.height=4}

## QC histograms
qc_tbl <- seu_tbl_full50 %>% 
  mutate(`log2(#UMIs)` = log2(nCount_RNA),
         `log2(#Genes)` = log2(nFeature_RNA),
         `log2(#UMIs/#Genes)` = `log2(#UMIs)`-`log2(#Genes)`,
         `MT reads [%]` = percent.mt) %>% 
  select(`log2(#UMIs)`, 
         `log2(#Genes)`, 
         # `log2(#UMIs/#Genes)`, 
         `MT reads [%]`, cell_type) %>% 
  gather(key, value, -cell_type) %>% 
  group_by(key) %>% 
  mutate(median_value = median(value))

qc_tbl <- bind_rows(qc_tbl, mutate(qc_tbl, cell_type = "All cell types")) %>% 
  mutate(cell_type = ordered(cell_type, 
                             levels = rev(c("All cell types", names(clrs$cell_type)))))

qc_plot <- ggplot(qc_tbl) +
  geom_violin(aes(cell_type, value, fill = cell_type), 
              color = "white", width = 1) +
  geom_boxplot(aes(cell_type, value, color = cell_type), 
               width = 0.15, size = 1, fill = NA, outlier.shape = NA) +
  geom_boxplot(aes(cell_type, value), fill = NA, 
               width = 0.15, color = "white", outlier.shape = NA) +
  geom_hline(aes(yintercept = median_value), 
             data = distinct(qc_tbl, key, median_value)) +
  facet_wrap(~key, scales = "free_x", strip.position = "bottom") +
  scale_fill_manual(values = c(clrs$cell_type, `All cell types` = "grey10"),
                    guide = F) +
  scale_color_manual(values = c(clrs$cell_type, `All cell types` = "grey10"),
                     guide = F) +
  coord_flip() +
  theme(axis.title = element_blank(),
        strip.placement = "outside")

```

```{r chunk_105, fig.width=16, fig.height=18}

mutsig_legend <- cowplot::get_legend(umap_mutsig + theme(legend.title=element_text()) + labs(color = "Mutational\nsignature"))
consOV_legend <- cowplot::get_legend(umap_consOV + theme(legend.title=element_text()) + labs(color = "TCGA subtype"))
cell_type_legend <- cowplot::get_legend(umap_cell_type + theme(legend.title=element_text(), legend.text = element_text(size = 14, margin = margin(0, 10, 0, 0), color = "black")) + labs(color = "Cell type"))
site_legend <- cowplot::get_legend(umap_site + theme(legend.title=element_text()) + labs(color = "Site"))

figS1_panel <- ggdraw() + 
  draw_plot(add_umap_coord(umap_mutsig + guides(color = F) + labs(title = "Mutational signature")), width = 0.24, height = 0.24, x = 0.01, y = 0.75) +
  draw_plot(add_umap_coord(umap_mitos), width = 0.24, height = 0.24, x = 0.25, y = 0.75) +
  draw_plot(add_umap_coord(umap_umis), width = 0.24, height = 0.24, x = 0.01, y = 0.5) +
  draw_plot(add_umap_coord(umap_genes), width = 0.24, height = 0.24, x = 0.25, y = 0.5) +
  draw_plot(qc_plot, width = 0.5, height = 0.25, x = 0.5, y = 0.75) +
  draw_plot(comp_grid_per_patient_p, width = 1, height = 0.23, x = 0, y = 0.25) +
  draw_plot(comp_grid_per_patient_n, width = 1, height = 0.23, x = 0, y = 0) +
  draw_grob(mutsig_legend, x = 0.55, y = 0.73, hjust = 0, vjust = 1) +
  draw_grob(consOV_legend, x = 0.55, y = 0.6, hjust = 0, vjust = 1) +
  draw_grob(cell_type_legend, x = 0.7, y = 0.73, hjust = 0, vjust = 1) + 
  draw_grob(site_legend, x = 0.85, y = 0.73, hjust = 0, vjust = 1) +
  draw_label("A", x = 0.01, y = 0.99, fontface = "bold", size = 22) +
  draw_label("B", x = 0.52, y = 0.99, fontface = "bold", size = 22) +
  draw_label("C", x = 0.52, y = 0.73, fontface = "bold", size = 22) +
  draw_label("D", x = 0.01, y = 0.47, fontface = "bold", size = 22) +
  draw_label("E", x = 0.01, y = 0.22, fontface = "bold", size = 22)

ggsave("_fig/001x_supplementary.pdf", figS1_panel, width = 16, height = 18)
figS1_panel

```

```{r chunk_106, fig.width=3, fig.height=5}

# matched_samples <- comp_tbl_sample_sort %>%
#   mutate(sample_id_x = str_remove_all(sample_id, " CD45-| CD45\\+")) %>%
#   distinct(sample_id_x, sample_id) %>%
#   group_by(sample_id_x) %>%
#   tally %>%
#   filter(n==2) %>%
#   pull(sample_id_x)
# 
# comp_rank_scatter_tbl <- comp_tbl_sample_sort %>%
#   mutate(sample_id_x = str_remove_all(sample_id, " CD45-| CD45\\+")) %>%
#   filter(sample_id_x %in% matched_samples,
#          (cell_type == "T cell" & sort_short_x == "CD45+") | (cell_type == "Ovarian cancer cell" & sort_short_x == "CD45-")) %>%
#   mutate(rank = ifelse(sort_short_x == "CD45-", rank_Ovarian.cancer.cell, rank_T.cell)) %>%
#   distinct(rank, sample_id_x, sort_short_x, tumor_supersite) %>%
#   spread(sort_short_x, rank) %>%
#   na.omit()
# 
# ggplot(comp_rank_scatter_tbl) +
#   geom_point(aes(`CD45-`, `CD45+`, color = tumor_supersite)) +
#   scale_color_manual(values = clrs$tumor_supersite)

```


## Mixing per cell type

```{r chunk_110, fig.width=3, fig.height=5}

pm_score_tbl <- read_tsv("/work/shah/uhlitzf/data/SPECTRUM/freeze/v5/patient_mixing_scores.tsv") %>% 
  mutate(cell_type = str_replace_all(cell_type, "Monocyte", "Myeloid cell"))

pm_score_tbl_ordered <- pm_score_tbl %>% 
  filter(cell_type != "Other") %>% 
  group_by(cell_type) %>% 
  mutate(median = median(score)) %>% 
  ungroup() %>% 
  arrange(-median) %>% 
  mutate(cell_type = ordered(cell_type, levels = unique(cell_type)))

mixing_plot_cell_type <- ggplot(pm_score_tbl_ordered) +
  geom_violin(aes(cell_type, score, fill = cell_type), 
              adjust = 3, color = "white", width = 1) +
  geom_boxplot(aes(cell_type, score, color = cell_type), 
               width = 0.2, size = 1, fill = NA, outlier.size = 0.01) +
  geom_boxplot(aes(cell_type, score, fill = cell_type), 
               width = 0.2, color = "white", outlier.shape = NA) +
  geom_hline(yintercept = 0, linetype = 2, size = 0.5, color = "black") +
  scale_fill_manual(values = clrs$cell_type) +
  scale_color_manual(values = clrs$cell_type) +
  guides(fill = F, color = F) +
  scale_y_continuous(expand = c(0, 0), 
                     limits = c(-5, 7),
                     breaks = c(-4, -2, 0, 2, 4, 6)) +  
  labs(y = "Patient specificity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank())

ggsave("_fig/_paper/01_patient_mixing_plot_v5.pdf", mixing_plot_cell_type,
       width = 3, height = 5)

```

### pairwise comparisons

```{r chunk_115, fig.width=3, fig.height=5}

pm_score_summary <- pm_score_tbl_ordered %>% 
  separate(cell_id, into = c("patient_id", "sample_suffix"), remove = F, extra = "drop", sep = "_") %>% 
  group_by(patient_id, cell_type) %>% 
  summarise(mean_score = mean(score))

ptt_tbl <- pairwise.t.test(pm_score_summary$mean_score, pm_score_summary$cell_type, p.adjust.method = "fdr") %$%
  p.value %>% 
  as_tibble(rownames = "cell_type_1") %>% 
  gather(cell_type_2, pvalue, -cell_type_1) %>% 
  mutate(cell_type_2 = ordered(cell_type_2, levels = levels(pm_score_tbl_ordered$cell_type))) %>% 
  mutate(cell_type_1 = ordered(cell_type_1, levels = levels(pm_score_tbl_ordered$cell_type))) %>% 
  mutate(log10p = -log10(pvalue),
         sig = pvalue < 0.1,
         label = ifelse(sig, "*", "")) %>% 
  na.omit()

pm_comparisons <- ggplot(filter(ptt_tbl, sig)) + 
  # geom_tile(aes(fct_rev(cell_type_1), fct_rev(cell_type_2), fill = log10p)) +
  geom_point(aes(fct_rev(cell_type_1), fct_rev(cell_type_2), size = log10p), color = "black", shape = 21) +
  geom_point(aes(fct_rev(cell_type_1), fct_rev(cell_type_2), fill = log10p, size = log10p), shape = 21) +
  # geom_point(aes(fct_rev(cell_type_1), fct_rev(cell_type_2), shape = sig),
  #           color = "white", size = 5) + 
  # geom_point(aes(fct_rev(cell_type_1), fct_rev(cell_type_2), shape = sig), 
  #           color = "black", size = 3) + 
  scale_fill_gradientn(colours = viridis(9)) +
  scale_size(guide = F) + 
  # scale_color_gradientn(colours = viridis(9)) +
  theme(aspect.ratio = 1, 
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title = element_blank(),
        plot.title = element_text(face = "plain", size = 12)) + 
  scale_shape_manual(values = c("", "*"), guide = F) +
  # scale_x_discrete(expand = c(0, 0)) +
  # scale_y_discrete(expand = c(0, 0)) +
  labs(fill = "-log10(p-value)", title = "Pairwise comparisons\nof patient specificity")

```

## consensusOV summary

```{r chunk_120, fig.width=6, fig.height=9}

consOV_tbl_summary <- seu_tbl_full50 %>%
  select(cell_type, Phase, S.Score, G2M.Score, consensusOV, patient_id_short, tumor_supersite) %>%
  mutate(consOV_cell_type = cell_type) %>% 
  # mutate(consOV_cell_type = ifelse(S.Score > 0.5, "Cycling cell", as.character(cell_type))) %>% 
  group_by(consOV_cell_type, consensusOV) %>%
  tally %>%
  group_by(consOV_cell_type) %>%
  mutate(nrel = n/(sum(n))*100) %>%
  ungroup %>% 
  mutate(consensusOV = ordered(consensusOV, levels = c(names(clrs$consensusOV)))) %>%
  arrange(consensusOV, nrel)

# consOV_tbl_summary <- seu_tbl_full50 %>%
#   select(cell_type, Phase, S.Score, G2M.Score, consensusOV, patient_id_short, tumor_supersite) %>%
#   mutate(consOV_cell_type = cell_type) %>%
#   # mutate(consOV_cell_type = ifelse(S.Score > 0.5, "Cycling cell", as.character(cell_type))) %>%
#   group_by(consOV_cell_type, consensusOV, tumor_supersite, patient_id_short) %>%
#   tally %>%
#   group_by(consOV_cell_type, tumor_supersite, patient_id_short) %>%
#   mutate(nrel = n/(sum(n))*100) %>%
#   ungroup %>%
#   mutate(consensusOV = ordered(consensusOV, levels = c(names(clrs$consensusOV)))) %>%
#   arrange(consensusOV, nrel) %>%
#   group_by(consOV_cell_type, tumor_supersite, consensusOV) %>%
#   summarise(mean_nrel = mean(nrel)) %>%
#   complete(tumor_supersite, nesting(consOV_cell_type, consensusOV), fill = list(mean_nrel = 0))
 
# consOV_tbl_summary <- seu_tbl_full50 %>%
#   select(cell_type, Phase, S.Score, G2M.Score, consensusOV, patient_id_short, tumor_supersite) %>%
#   mutate(consOV_cell_type = cell_type) %>% 
#   # mutate(consOV_cell_type = ifelse(S.Score > 0.5, "Cycling cell", as.character(cell_type))) %>% 
#   group_by(consOV_cell_type, consensusOV, tumor_supersite, patient_id_short) %>%
#   tally %>%
#   group_by(consOV_cell_type, tumor_supersite, patient_id_short) %>%
#   mutate(nrel = n/(sum(n))*100) %>%
#   ungroup %>% 
#   mutate(consensusOV = ordered(consensusOV, levels = c(names(clrs$consensusOV)))) %>%
#   arrange(consensusOV, nrel) %>%
#   ungroup() %>% 
#   complete(tumor_supersite, nesting(consOV_cell_type, consensusOV, patient_id_short), fill = list(nrel = 0)) %>% 
#   complete(patient_id_short, nesting(consOV_cell_type, consensusOV, tumor_supersite), fill = list(nrel = 0))

heat_layers <- list(
  geom_tile(),
  scale_fill_gradientn(colors = magma(9)[c(-1)], limits = c(0, 100), breaks = c(0, 50, 100)),
  theme(
    axis.title = element_blank(),
    axis.ticks = element_blank(),
    axis.line = element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    legend.position = "bottom",
    aspect.ratio = 9/4),
  guides(fill = guide_colorbar(title.position = "top"))
)

tcga_heat <- consOV_tbl_summary %>% 
  filter(consOV_cell_type != "Other") %>% 
  mutate(cell_type = ordered(consOV_cell_type, levels = unique(c("Myeloid cell", "Fibroblast", "Cycling cell", names(clrs$cell_type))))) %>% 
  mutate(consensusOV = ordered(consensusOV, levels = names(clrs$consensusOV))) %>% 
  ggplot(aes(consensusOV, cell_type, fill = nrel)) +
  labs(fill = "Fraction\nof cells [%]") +
  heat_layers

# tcga_heat + guides(fill = F)
tcga_heat_legend <- get_legend(tcga_heat)

ggsave("_fig/001x_consOV_heat.pdf", tcga_heat, width = 25, height = 30)

```

## summary stats grid

```{r chunk_130, fig.width=6, fig.height=10}

right_grid_full <- ggdraw() +
  draw_plot(tcga_heat + guides(fill = F), x = -0.05, y = 0.58, width = 0.65, height = 0.4) +
  draw_grob(tcga_heat_legend, x = 0.05, y = 0.56, width = 0.2, height = 0.2) +
  draw_plot(mixing_plot_cell_type, x = 0.56, y = 0.58, width = 0.45, height = 0.4) +
  draw_plot(pcomp_grid_p_full, x = 0, y = 0, width = 0.5, height = 0.55) +
  draw_plot(pcomp_grid_n_full, x = 0.5, y = 0, width = 0.5, height = 0.55)
  
right_grid_full

ggsave("_fig/001x_right_grid_full.png", right_grid_full, width = 6, height = 10)
ggsave("_fig/001x_right_grid_full.pdf", right_grid_full, width = 6, height = 10)

```

```{r chunk_135, fig.width=6, fig.height=9}

right_grid <- ggdraw() +
  draw_plot(tcga_heat + guides(fill = F), x = -0.05, y = 0.65, width = 0.65, height = 0.35) +
  draw_grob(tcga_heat_legend, x = 0.05, y = 0.6, width = 0.2, height = 0.2) +
  draw_plot(mixing_plot_cell_type, x = 0.56, y = 0.65, width = 0.45, height = 0.35) +
  draw_plot(pcomp_grid_p, x = 0, y = 0, width = 0.5, height = 0.6) +
  draw_plot(pcomp_grid_n, x = 0.5, y = 0, width = 0.5, height = 0.6)
  
right_grid

ggsave("_fig/001x_right_grid.png", right_grid, width = 6, height = 10)
ggsave("_fig/001x_right_grid.pdf", right_grid, width = 6, height = 10)

```

## mpIF composition

```{r chunk_140, fig.width=6, fig.height=3}

## meta data for mpIF
mpif_meta_tbl <- read_excel("_data/small/MSK SPECTRUM - mpIF.xlsx", sheet = 3) %>% 
  mutate(slide_id = str_replace_all(pici_id, " ", "_"),
         sample_cd45p = paste0(patient_id, "_", surgery, "_CD45P_", str_replace_all(toupper(tumor_subsite), " ", "_")),
         sample_cd45n = paste0(patient_id, "_", surgery, "_CD45N_", str_replace_all(toupper(tumor_subsite), " ", "_"))) %>% 
  mutate(tumor_supersite = ifelse(tumor_supersite == "Upper Quadrant", "UQ", tumor_supersite))

# mpif_pixel <- read_tsv("/work/shah/vazquezi/data/transfers/spectrum/results/mpif/v8/integrate/outputs/cohort_merge/patient/SPECTRUM/all/detection.tsv") %>% 
#   select(cell_id, compartment = Parent)
# 
# mpif_cell_type <- read_tsv("/work/shah/vazquezi/data/transfers/spectrum/results/mpif/v8/integrate/outputs/cohort_merge/patient/SPECTRUM/all/cell_type_manual.tsv") %>%
#   left_join(mpif_pixel, by = "cell_id") %>% 
#   mutate(cell_id = str_remove_all(cell_id, "CD68.TOX.PD1.PDL1.CD8.panCK_CK8-18.DAPI_|_component_data - resolution #1")) %>%
#   separate(cell_id, into = c("patient_id", "tumor_subsite", "fov_id", "cell_idx"),
#            sep = "_", remove = F) %>%
#   mutate(fov_id = paste0(patient_id, "_", tumor_subsite, "_", fov_id),
#          slide_id = paste0(patient_id, "_", tumor_subsite))
# 
# ## expand for double and triple positive cells (cell type markers)
# mpif_ncell <- mpif_cell_type %>%
#   # sample_n(10000) %>%
#   mutate(CD68_state_log = CD68_state == "CD68+",
#          CD8_state_log = CD8_state == "CD8+",
#          panCK_state_log = panCK_state == "panCK+") %>%
#   group_by(cell_id) %>%
#   mutate(n_cells = sum(CD68_state_log, CD8_state_log, panCK_state_log)) %>%
#   select(cell_id, n_cells) %>%
#   deframe
# 
# mpif_cell_type_expanded <- bind_rows(
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 0]),
#          TOX_state == "TOX+" | PD1_state == "PD1+" | PDL1_state == "PDL1+") %>%
#     mutate(cell_id = paste0(cell_id, "_0"),
#            cell_type = c("Other")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 0]),
#          TOX_state == "TOX-" & PD1_state == "PD1-" & PDL1_state == "PDL1-") %>%
#     mutate(cell_id = paste0(cell_id, "_0"),
#            cell_type = c("Unknown")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 1]),
#          CD8_state == "CD8+") %>%
#     mutate(cell_id = paste0(cell_id, "_1"),
#            cell_type = c("CD8+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 1]),
#          CD68_state == "CD68+") %>%
#     mutate(cell_id = paste0(cell_id, "_2"),
#            cell_type = c("CD68+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 1]),
#          panCK_state == "panCK+") %>%
#     mutate(cell_id = paste0(cell_id, "_3"),
#            cell_type = c("panCK+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 2]),
#          CD8_state == "CD8+") %>%
#     mutate(cell_id = paste0(cell_id, "_1"),
#            cell_type = c("CD8+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 2]),
#          CD68_state == "CD68+") %>%
#     mutate(cell_id = paste0(cell_id, "_2"),
#            cell_type = c("CD68+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 2]),
#          panCK_state == "panCK+") %>%
#     mutate(cell_id = paste0(cell_id, "_3"),
#            cell_type = c("panCK+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 3]),
#          CD8_state == "CD8+") %>%
#     mutate(cell_id = paste0(cell_id, "_1"),
#            cell_type = c("CD8+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 3]),
#          CD68_state == "CD68+") %>%
#     mutate(cell_id = paste0(cell_id, "_2"),
#            cell_type = c("CD68+")),
#   filter(mpif_cell_type, cell_id %in% names(mpif_ncell[mpif_ncell == 3]),
#          panCK_state == "panCK+") %>%
#     mutate(cell_id = paste0(cell_id, "_3"),
#            cell_type = c("panCK+"))
# )
# 
# mpif_cell_state_expanded <- mpif_cell_type_expanded %>%
#   mutate(cell_state = case_when(
#     cell_type == "CD8+" ~ paste0(CD8_state, TOX_state, PD1_state),
#     cell_type == "CD68+" ~ paste0(CD68_state, PDL1_state),
#     cell_type == "panCK+" ~ paste0(panCK_state, PDL1_state),
#     cell_type == "Unknown" ~ "Unknown",
#     cell_type == "Other" ~ "Other"
#   ))
# 
# write_tsv(mpif_cell_state_expanded, "/work/shah/uhlitzf/data/SPECTRUM/mpIF/cell_type_manual_expanded.tsv")

mpif_cell_state_expanded <- read_tsv("/work/shah/uhlitzf/data/SPECTRUM/mpIF/cell_type_manual_expanded.tsv") %>% 
  select(cell_id, slide_id, fov_id, compartment, cell_type, cell_state, contains("state")) %>% 
  left_join(select(mpif_meta_tbl, slide_id, patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, sample_cd45n), by = "slide_id") %>% 
  na.omit()

# ## cell type composition 
# mpif_cell_type_n <- mpif_cell_state_expanded %>%
#   group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, fov_id, cell_type) %>%
#   tally() %>%
#   group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, fov_id) %>%
#   mutate(nrel = n/sum(n)) %>%
#   ungroup()
# 
# ## cell type composition slide lvl
# mpif_cell_type_n_slide <- mpif_cell_state_expanded %>%
#   group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, cell_type) %>%
#   tally() %>%
#   group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p) %>%
#   mutate(nrel = n/sum(n)) %>%
#   ungroup()

## cell state composition fov lvl
mpif_cell_state_n <- mpif_cell_state_expanded %>%
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, fov_id, cell_state, cell_type) %>%
  tally() %>%
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, fov_id) %>%
  mutate(nrel = n/sum(n)) %>%
  ungroup()

## cell state composition slide lvl
mpif_cell_state_n_slide <- mpif_cell_state_expanded %>%
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, cell_state, cell_type) %>%
  tally() %>%
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p) %>%
  mutate(nrel = n/sum(n)) %>%
  ungroup()

## cell state composition slide lvl compartment
mpif_cell_state_n_slide_compartment <- mpif_cell_state_expanded %>%
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, compartment, cell_state, cell_type) %>%
  tally() %>%
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, compartment) %>%
  mutate(nrel = n/sum(n)) %>%
  ungroup()

## fov tally helper function
fov_tally <- . %>% 
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, fov_id) %>%
  mutate(nrel = n/sum(n)) %>% 
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, fov_id, cell_type) %>% 
  mutate(nrel_ct = sum(nrel)) %>% 
  ungroup

## slide tally helper function
slide_tally <- . %>% 
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p) %>%
  mutate(nrel = n/sum(n)) %>% 
  group_by(patient_id, tumor_supersite, tumor_subsite, therapy, sample_cd45p, cell_type) %>% 
  mutate(nrel_ct = sum(nrel)) %>% 
  ungroup() %>% 
  arrange(cell_type != "CD8+", desc(cell_type), nrel_ct) %>% 
  mutate(sample_cd45p = ordered(sample_cd45p, levels = unique(sample_cd45p)),
         sort_short_x = "mpIF",
         label_supersite = "Site",
         label_therapy = "Rx",
         label_mutsig = "Signature",
         patient_id_short = str_sub(patient_id, 13, 15)) %>% 
  distinct(sample_cd45p, cell_state, .keep_all = T) %>% 
  arrange(sort_short_x, cell_type, cell_state) %>% 
  group_by(cell_type, cell_state, sort_short_x) %>% 
  mutate(rank_T.cell = row_number(sample_cd45p)) %>% 
  mutate(rank_T.cell = scales::rescale(rank_T.cell, c(-1, 1))) %>% 
  ungroup %>% 
  mutate(tumor_supersite = ordered(tumor_supersite, levels = rev(names(clrs$tumor_supersite))),
         cell_state = ordered(cell_state, levels = names(clrs$cell_state)))

## immune state composition fov lvl
mpif_cell_state_n_imm <- mpif_cell_state_n %>% 
  filter(cell_type %in% c("CD68+", "CD8+")) %>% 
  fov_tally

## immune state composition slide lvl
mpif_cell_state_n_imm_slide <- mpif_cell_state_n_slide %>% 
  # filter(cell_type %in% c("CD68+", "CD8+")) %>% 
  slide_tally

## immune state composition slide lvl per tumor
mpif_cell_state_n_imm_slide_tumor <- mpif_cell_state_n_slide_compartment %>% 
  filter(compartment == "Tumor", therapy == "pre-Rx") %>% 
  slide_tally %>% 
  mutate(sort_short_x = compartment)

## immune state composition slide lvl per stroma
mpif_cell_state_n_imm_slide_stroma <- mpif_cell_state_n_slide_compartment %>% 
  filter(compartment == "Stroma", therapy == "pre-Rx") %>% 
  slide_tally %>% 
  mutate(sort_short_x = compartment)

## CD68 state composition
mpif_cell_state_n_imm_cd68 <- mpif_cell_state_n %>% 
  filter(cell_type %in% c("CD68+")) %>% 
  fov_tally

## CD68 state composition slide lvl
mpif_cell_state_n_imm_cd68_slide <- mpif_cell_state_n_slide %>% 
  filter(cell_type %in% c("CD68+")) %>% 
  slide_tally

## CD8 state composition
mpif_cell_state_n_imm_cd8 <- mpif_cell_state_n %>% 
  filter(cell_type %in% c("CD8+")) %>% 
  fov_tally

## CD8 state composition slide lvl
mpif_cell_state_n_imm_cd8_slide <- mpif_cell_state_n_slide %>% 
  filter(cell_type %in% c("CD8+")) %>% 
  slide_tally

# mpIF_cell_clrs <- c(clrs$cell_type, clrs$cluster_label$T.cell)
# names(mpIF_cell_clrs) <- str_replace_all(names(mpIF_cell_clrs), " ", ".")

```

## mpif vector ranks 

```{r chunk_150, fig.width=6, fig.height=3}

prim_ranks_cd45p_mpif <- filter(mpif_cell_state_n_imm_slide) %>%
  filter((tumor_supersite %in% c("Adnexa")), cell_type == "CD8+", cell_state == "CD8+TOX-PD1-") %>%
  mutate(rank = rank_T.cell) %>%
  prim_ranks_wrapper("Adnexa")

meta_ranks_cd45p_mpif <- filter(mpif_cell_state_n_imm_slide) %>%
  filter(!(tumor_supersite %in% c("Adnexa")), cell_type == "CD8+", cell_state == "CD8+TOX-PD1-") %>%
  mutate(rank = rank_T.cell) %>%
  prim_ranks_wrapper("Non-adnexa")

median_rank_tbl_cd45p_mpif <- bind_rows(distinct(prim_ranks_cd45p_mpif, median_rank, .keep_all = T),
                                        distinct(meta_ranks_cd45p_mpif, median_rank, .keep_all = T)) %>%
  median_rank_wrapper

rank_tbl_cd45p_mpif <- bind_rows(prim_ranks_cd45p_mpif, meta_ranks_cd45p_mpif) %>%
  mutate(prim_meta = ifelse(tumor_supersite == "Ascites", "Ascites", prim_meta),
         patient_id_short = ordered(patient_id_short, levels = levels(median_rank_tbl_cd45p_mpif$patient_id_short)))

vector_patient_lvls_cd45p_mpif <- levels(rank_tbl_cd45p$patient_id_short)
vector_patient_lvls_cd45p_mpif[as.logical(1:length(vector_patient_lvls_cd45p_mpif) %% 2)] <- paste0(vector_patient_lvls_cd45p_mpif[as.logical(1:length(vector_patient_lvls_cd45p_mpif) %% 2)], "       ")

vector_rank_list_cd45p_mpif <- list(rank_tbl = rank_tbl_cd45p_mpif,
                               median_rank_tbl = median_rank_tbl_cd45p_mpif,
                               vector_patient_lvls = vector_patient_lvls_cd45p_mpif)

vector_plot_cd45p_mpif <- vector_patient_plot_wrapper(vector_rank_list_cd45p_mpif, c("#ff7f00", "#33a02c"))


## tumor compartment vectors
prim_ranks_tumor_mpif <- filter(mpif_cell_state_n_imm_slide_tumor, sort_short_x == "Tumor", therapy == "pre-Rx", compartment == "Tumor") %>%
  filter((tumor_supersite %in% c("Adnexa")), cell_type == "CD8+", cell_state == "CD8+TOX-PD1-") %>%
  mutate(rank = rank_T.cell) %>%
  prim_ranks_wrapper("Adnexa")

meta_ranks_tumor_mpif <- filter(mpif_cell_state_n_imm_slide_tumor, sort_short_x == "Tumor", therapy == "pre-Rx", compartment == "Tumor") %>%
  filter(!(tumor_supersite %in% c("Adnexa")), cell_type == "CD8+", cell_state == "CD8+TOX-PD1-") %>%
  mutate(rank = rank_T.cell) %>%
  prim_ranks_wrapper("Non-adnexa")

median_rank_tbl_tumor_mpif <- bind_rows(distinct(prim_ranks_tumor_mpif, median_rank, .keep_all = T),
                                        distinct(meta_ranks_tumor_mpif, median_rank, .keep_all = T)) %>%
  median_rank_wrapper

rank_tbl_tumor_mpif <- bind_rows(prim_ranks_tumor_mpif, meta_ranks_tumor_mpif) %>%
  mutate(prim_meta = ifelse(tumor_supersite == "Ascites", "Ascites", prim_meta),
         patient_id_short = ordered(patient_id_short, levels = levels(median_rank_tbl_tumor_mpif$patient_id_short)))

vector_patient_lvls_tumor_mpif <- levels(rank_tbl_cd45p$patient_id_short)
vector_patient_lvls_tumor_mpif[as.logical(1:length(vector_patient_lvls_tumor_mpif) %% 2)] <- paste0(vector_patient_lvls_tumor_mpif[as.logical(1:length(vector_patient_lvls_tumor_mpif) %% 2)], "       ")

vector_rank_list_tumor_mpif <- list(rank_tbl = rank_tbl_tumor_mpif,
                               median_rank_tbl = median_rank_tbl_tumor_mpif,
                               vector_patient_lvls = vector_patient_lvls_tumor_mpif)

vector_plot_tumor_mpif <- vector_patient_plot_wrapper(vector_rank_list_tumor_mpif, c("#ff7f00", "#33a02c"))


## stroma compartment vectors
prim_ranks_stroma_mpif <- filter(mpif_cell_state_n_imm_slide_stroma, sort_short_x == "Stroma", therapy == "pre-Rx", compartment == "Stroma") %>%
  filter((tumor_supersite %in% c("Adnexa")), cell_type == "CD8+", cell_state == "CD8+TOX-PD1-") %>%
  mutate(rank = rank_T.cell) %>%
  prim_ranks_wrapper("Adnexa")

meta_ranks_stroma_mpif <- filter(mpif_cell_state_n_imm_slide_stroma, sort_short_x == "Stroma", therapy == "pre-Rx", compartment == "Stroma") %>%
  filter(!(tumor_supersite %in% c("Adnexa")), cell_type == "CD8+", cell_state == "CD8+TOX-PD1-") %>%
  mutate(rank = rank_T.cell) %>%
  prim_ranks_wrapper("Non-adnexa")

median_rank_tbl_stroma_mpif <- bind_rows(distinct(prim_ranks_stroma_mpif, median_rank, .keep_all = T),
                                        distinct(meta_ranks_stroma_mpif, median_rank, .keep_all = T)) %>%
  median_rank_wrapper

rank_tbl_stroma_mpif <- bind_rows(prim_ranks_stroma_mpif, meta_ranks_stroma_mpif) %>%
  mutate(prim_meta = ifelse(tumor_supersite == "Ascites", "Ascites", prim_meta),
         patient_id_short = ordered(patient_id_short, levels = levels(median_rank_tbl_stroma_mpif$patient_id_short)))

vector_patient_lvls_stroma_mpif <- levels(rank_tbl_cd45p$patient_id_short)
vector_patient_lvls_stroma_mpif[as.logical(1:length(vector_patient_lvls_stroma_mpif) %% 2)] <- paste0(vector_patient_lvls_stroma_mpif[as.logical(1:length(vector_patient_lvls_stroma_mpif) %% 2)], "       ")

vector_rank_list_stroma_mpif <- list(rank_tbl = rank_tbl_stroma_mpif,
                               median_rank_tbl = median_rank_tbl_stroma_mpif,
                               vector_patient_lvls = vector_patient_lvls_stroma_mpif)

vector_plot_stroma_mpif <- vector_patient_plot_wrapper(vector_rank_list_stroma_mpif, c("#ff7f00", "#33a02c"))


```

```{r chunk_160, fig.width=3, fig.height=6}

pcomp1 <- comp_plot_wrapper(mpif_cell_state_n_imm_slide, "mpIF", "n", x = "sample_cd45p", facet = F, fill = "cell_state") +
  scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 250000, 500000),
                       limits = c(0, 500000),
                       labels = c("", "250000", "500000")) +
    expand_limits(y = c(0, 500000))

pcomp2 <- comp_plot_wrapper(mpif_cell_state_n_imm_slide, "mpIF", "nrel", x = "sample_cd45p", facet = F, fill = "cell_state") +
  scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 0.5, 1),
                       limits = c(0, 1),
                       labels = c("0", "50", "100")) +
  expand_limits(y = c(0, 1))


pcomp_grid_p_mpif <- plot_grid(pcomp1, pcomp2,
                               comp_label_boxplot(filter(mpif_cell_state_n_imm_slide, cell_state == "CD8+TOX-PD1-"), x = "tumor_supersite", y = "rank_T.cell", facet = F, cts = "mpIF", ct = "CD8+", pvals = F) + remove_xaxis, 
                               vector_plot_cd45p + remove_guides, 
                               ncol = 1, align = "v", 
                               rel_heights = c(0.15, 0.15, 0.25, 0.45))

ggsave("_fig/001x_sorted_comp_mpif.pdf", pcomp_grid_p_mpif, width = 3, height = 6)


```

```{r chunk_161, fig.width=3, fig.height=6}

pcomp1 <- comp_plot_wrapper(mpif_cell_state_n_imm_slide_tumor, "Tumor", "n", x = "sample_cd45p", facet = F, fill = "cell_state") +
  scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 150000, 300000),
                       limits = c(0, 300000),
                       labels = c("", "150000", "300000")) +
    expand_limits(y = c(0, 300000))

pcomp2 <- comp_plot_wrapper(mpif_cell_state_n_imm_slide_tumor, "Tumor", "nrel", x = "sample_cd45p", facet = F, fill = "cell_state") +
  scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 0.5, 1),
                       limits = c(0, 1),
                       labels = c("0", "50", "100")) +
  expand_limits(y = c(0, 1))


pcomp_grid_p_mpif_tumor <- plot_grid(pcomp1, pcomp2,
                               comp_label_boxplot(filter(mpif_cell_state_n_imm_slide_tumor, cell_state == "CD8+TOX-PD1-"), x = "tumor_supersite", y = "rank_T.cell", facet = F, cts = "Tumor", ct = "CD8+", pvals = F) + remove_xaxis, 
                               vector_plot_tumor_mpif + remove_guides, 
                               ncol = 1, align = "v", 
                               rel_heights = c(0.15, 0.15, 0.25, 0.45))

ggsave("_fig/001x_sorted_comp_mpif_tumor.pdf", pcomp_grid_p_mpif, width = 3, height = 6)


```

```{r chunk_162, fig.width=3, fig.height=6}

pcomp1 <- comp_plot_wrapper(mpif_cell_state_n_imm_slide_stroma, "Stroma", "n", x = "sample_cd45p", facet = F, fill = "cell_state") +
  scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 150000, 300000),
                       limits = c(0, 300000),
                       labels = c("", "150000", "300000")) +
    expand_limits(y = c(0, 300000))

pcomp2 <- comp_plot_wrapper(mpif_cell_state_n_imm_slide_stroma, "Stroma", "nrel", x = "sample_cd45p", facet = F, fill = "cell_state") +
  scale_y_continuous(expand = c(0, 0), 
                       breaks = c(0, 0.5, 1),
                       limits = c(0, 1),
                       labels = c("0", "50", "100")) +
  expand_limits(y = c(0, 1))


pcomp_grid_p_mpif_stroma <- plot_grid(pcomp1, pcomp2,
                               comp_label_boxplot(filter(mpif_cell_state_n_imm_slide_stroma, cell_state == "CD8+TOX-PD1-"), x = "tumor_supersite", y = "rank_T.cell", facet = F, cts = "Stroma", ct = "CD8+", pvals = F) + remove_xaxis, 
                               vector_plot_stroma_mpif + remove_guides, 
                               ncol = 1, align = "v", 
                               rel_heights = c(0.15, 0.15, 0.25, 0.45))

ggsave("_fig/001x_sorted_comp_mpif_stroma.pdf", pcomp_grid_p_mpif, width = 3, height = 6)


```

## summary stats mpif comparison

```{r chunk_165, fig.width=12, fig.height=10}

right_grid <- ggdraw() +
  draw_plot(tcga_heat + guides(fill = F), x = -0.05, y = 0.65, width = 0.33, height = 0.35) +
  draw_grob(tcga_heat_legend, x = 0.05, y = 0.6, width = 0.1, height = 0.2) +
  draw_plot(mixing_plot_cell_type, x = 0.3, y = 0.65, width = 0.25, height = 0.35) +
  draw_plot(pcomp_grid_n, x = 0, y = 0, width = 0.25, height = 0.6) +
  draw_plot(pcomp_grid_p, x = 0.25, y = 0, width = 0.25, height = 0.6) +
  draw_plot(pcomp_grid_p_mpif_tumor, x = 0.5, y = 0, width = 0.25, height = 0.6) +
  draw_plot(pcomp_grid_p_mpif_stroma, x = 0.75, y = 0, width = 0.25, height = 0.6)
  
right_grid

ggsave("_fig/001x_right_grid_mpif.png", right_grid, width = 12, height = 10)
ggsave("_fig/001x_right_grid_mpif.pdf", right_grid, width = 12, height = 10)

```



## scRNA marker positivity

```{r chunk_170, fig.width=6, fig.height=3}

# seu_cohort <- read_rds("/work/shah/isabl_data_lake/analyses/16/52/1652/cohort_merged.rdata")
# 
# scrna_markers <- as_tibble(cbind(cell_id = colnames(seu_cohort), FetchData(seu_cohort, c("cell_type", "sample", "CD68", "PDCD1", "CD274", "TOX", "CD8A", "CD8B", "KRT8", "KRT19")))) %>%
#   mutate(CD68_state = ifelse(CD68 > 0, "CD68+", "CD68-"),
#          CD8_state = ifelse(CD8A > 0 | CD8B > 0, "CD8+", "CD8-"),
#          panCK_state = ifelse(KRT8 > 0 | KRT19 > 0, "panCK+", "panCK-"),
#          TOX_state = ifelse(TOX > 0, "TOX+", "TOX-"),
#          PD1_state = ifelse(PDCD1 > 0, "PD1+", "PD1-"),
#          PDL1_state = ifelse(CD274 > 0, "PDL1+", "PDL1-"))
# 
# write_tsv(scrna_markers, "/work/shah/uhlitzf/data/SPECTRUM/freeze/v5/scrna_mpif_marker_expression.tsv")

scrna_markers <- read_tsv("/work/shah/uhlitzf/data/SPECTRUM/freeze/v5/scrna_mpif_marker_expression.tsv") %>%
  filter(!(CD68_state == "CD68+" & CD8_state == "CD8+"),
         !(CD68_state == "CD68+" & panCK_state == "panCK+"),
         !(CD8_state == "CD8+" & panCK_state == "panCK+")) %>%
  mutate(cell_type_sc = ifelse(CD68_state == "CD68+", "CD68+", ifelse(CD8_state == "CD8+", "CD8+", ifelse(panCK_state == "panCK+", "panCK+", "Other"))),
         cell_state = case_when(
           cell_type_sc == "CD8+" ~ paste0(CD8_state, TOX_state, PD1_state),
           cell_type_sc == "CD68+" ~ paste0(CD68_state, PDL1_state),
           cell_type_sc == "panCK+" ~ paste0(panCK_state, PDL1_state),
           cell_type_sc == "Unknown" ~ "Unknown",
           cell_type_sc == "Other" ~ "Other"
         ))

markers_pos_frac_scrna_celltype <- seu_tbl_full50 %>%
  # select(cell_id, sample) %>%
  select(cell_id, sample, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>%
  left_join(select(scrna_markers, cell_id, cell_type_sc, cell_state), by = "cell_id") %>% 
  na.omit() %>% 
  group_by(sample, cell_type_sc, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>% 
  tally %>% 
  group_by(sample, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>% 
  mutate(nrel = n/sum(n)) %>% 
  select(sample, cell_type_sc, n, nrel, everything()) %>% 
  ungroup

markers_pos_frac_scrna_cellstate <- seu_tbl_full50 %>%
  # select(cell_id, sample) %>%
  select(cell_id, sample, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>%
  left_join(select(scrna_markers, cell_id, cell_type_sc, cell_state), by = "cell_id") %>% 
  na.omit() %>% 
  group_by(sample, cell_state, cell_type_sc, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>% 
  tally %>% 
  group_by(sample, cell_type_sc, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>% 
  mutate(nrel = n/sum(n)) %>% 
  select(sample, cell_type_sc, cell_state, n, nrel, everything()) %>% 
  ungroup

markers_pos_frac_scrna_gene <- seu_tbl_full50 %>%
  # select(cell_id, sample) %>%
  select(cell_id, sample, cell_type, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>%
  left_join(select(scrna_markers, -sample, -cell_type), by = "cell_id") %>% 
  na.omit() %>% 
  select(-contains("state")) %>% 
  mutate(cell_type = str_replace_all(cell_type, "\\.", " ")) %>%
  gather(gene, value, -cell_id, -sample, -cell_type, -tumor_supersite, 
         -consensus_signature, -patient_id_short, -sort_short_x, -cell_type_sc) %>% 
  mutate(gene_state = value > 0) %>% 
  group_by(gene, sample, tumor_supersite, consensus_signature, patient_id_short, sort_short_x) %>% 
  mutate(n = sum(gene_state),
         nrel = n/length(n)) %>% 
  ungroup %>% 
  distinct(gene, sample, tumor_supersite, consensus_signature, patient_id_short, sort_short_x, n, nrel) %>% 
  select(sample, gene, n, nrel, everything())
  

```

## mpIF x scRNA correlation

### CD45+ cell type correlation

```{r chunk_180, fig.width=6, fig.height=3}

markers_pos_scrna_mpif <- mpif_cell_state_n_slide_compartment %>% 
  select(sample = sample_cd45p, cell_type_sc = cell_type, 
         compartment, n_mpif = n, nrel_mpif = nrel) %>% 
  group_by(sample, cell_type_sc, compartment) %>% 
  summarise(n_mpif = sum(n_mpif), nrel_mpif = sum(nrel_mpif)) %>% 
  ungroup() %>% 
  left_join(markers_pos_frac_scrna_celltype, by = c("sample", "cell_type_sc")) %>% 
  filter(sort_short_x == "CD45+",
         cell_type_sc %in% c("CD8+", "CD68+"),
         compartment %in% c("Stroma", "Tumor"))

common_layers <- list(
  facet_wrap(cell_type_sc~compartment, scales = "free"),
  stat_smooth(aes(nrel, nrel_mpif), method = "lm", color = "black"),
  stat_cor(aes(nrel, nrel_mpif), method = "spearman", color = "black"),
  labs(x = "Fraction in scRNA",
       y = "Fraction in mpIF")
)

p1 <- ggplot(markers_pos_scrna_mpif) +
  geom_point(aes(nrel, nrel_mpif, color = tumor_supersite)) +
  scale_color_manual(values = clrs$tumor_supersite) +
  common_layers

p2 <- ggplot(markers_pos_scrna_mpif) +
  geom_point(aes(nrel, nrel_mpif, color = consensus_signature)) +
  scale_color_manual(values = clrs$consensus_signature) +
  common_layers

plot_grid(p1, p2)

```

### CD45+ cell state correlation

```{r chunk_190, fig.width=6, fig.height=3}

markers_pos_scrna_mpif_state <- mpif_cell_state_n_slide_compartment %>% 
  group_by(sample_cd45p, compartment, cell_type) %>% 
  mutate(nrel_state = n/sum(n)) %>% 
  select(sample = sample_cd45p, cell_type_sc = cell_type, 
         compartment, n_mpif = n, nrel_mpif = nrel_state, cell_state) %>% 
  left_join(markers_pos_frac_scrna_cellstate, by = c("sample", "cell_type_sc", "cell_state")) %>% 
  filter(sort_short_x == "CD45+",
         cell_state %in% c("CD8+TOX-PD1-", "CD8+TOX+PD1+", "CD68+PDL1-", "CD68+PDL1+"),
         compartment %in% c("Stroma", "Tumor"))

common_layers <- list(
  facet_wrap(cell_state~compartment, scales = "free", ncol = 4),
  geom_smooth(aes(nrel, nrel_mpif), method = "lm", color = "black"),
  stat_cor(aes(nrel, nrel_mpif), method = "spearman", color = "black"),
  labs(x = "Fraction in scRNA",
       y = "Fraction in mpIF")
)

p1 <- ggplot(markers_pos_scrna_mpif_state) +
  geom_point(aes(nrel, nrel_mpif, color = tumor_supersite)) +
  scale_color_manual(values = clrs$tumor_supersite) +
  common_layers

p2 <- ggplot(markers_pos_scrna_mpif_state) +
  geom_point(aes(nrel, nrel_mpif, color = consensus_signature)) +
  scale_color_manual(values = clrs$consensus_signature) +
  common_layers

plot_grid(p1, p2, ncol = 1)


```

### CD45- cell type correlation

```{r chunk_200, fig.width=6, fig.height=3}

markers_pos_scrna_mpif_cd45n <- mpif_cell_state_n_slide_compartment %>% 
  select(sample = sample_cd45p, cell_type_sc = cell_type, 
         compartment, n_mpif = n, nrel_mpif = nrel) %>% 
  mutate(sample = str_replace_all(sample, "CD45P", "CD45N")) %>% 
  group_by(sample, cell_type_sc, compartment) %>% 
  summarise(n_mpif = sum(n_mpif), nrel_mpif = sum(nrel_mpif)) %>% 
  ungroup() %>% 
  left_join(markers_pos_frac_scrna_celltype, by = c("sample", "cell_type_sc")) %>% 
  filter(sort_short_x == "CD45-",
         cell_type_sc %in% c("panCK+"),
         compartment %in% c("Stroma", "Tumor"))

common_layers <- list(
  facet_wrap(cell_type_sc~compartment, scales = "free"),
  stat_smooth(aes(nrel, nrel_mpif), method = "lm", color = "black"),
  stat_cor(aes(nrel, nrel_mpif), method = "pearson", color = "black"),
  labs(x = "Fraction in scRNA",
       y = "Fraction in mpIF")
)

p1 <- ggplot(markers_pos_scrna_mpif_cd45n) +
  geom_point(aes(nrel, nrel_mpif, color = tumor_supersite)) +
  scale_color_manual(values = clrs$tumor_supersite) +
  common_layers

p2 <- ggplot(markers_pos_scrna_mpif_cd45n) +
  geom_point(aes(nrel, nrel_mpif, color = consensus_signature)) +
  scale_color_manual(values = clrs$consensus_signature) +
  common_layers

plot_grid(p1, p2)


```

### CD45- cell state correlation

```{r chunk_210, fig.width=6, fig.height=3}

markers_pos_scrna_mpif_state_cd45n <- mpif_cell_state_n_slide_compartment %>% 
  group_by(sample_cd45p, compartment, cell_type) %>% 
  mutate(nrel_state = n/sum(n)) %>% 
  select(sample = sample_cd45p, cell_type_sc = cell_type, 
         compartment, n_mpif = n, nrel_mpif = nrel_state, cell_state) %>% 
  mutate(sample = str_replace_all(sample, "CD45P", "CD45N")) %>% 
  left_join(markers_pos_frac_scrna_cellstate, by = c("sample", "cell_type_sc", "cell_state")) %>% 
  filter(sort_short_x == "CD45-",
         cell_state %in% c("panCK+PDL1-", "panCK+PDL1+"),
         compartment %in% c("Stroma", "Tumor"))

common_layers <- list(
  facet_wrap(cell_state~compartment, scales = "free", ncol = 4),
  geom_smooth(aes(nrel, nrel_mpif), method = "lm", color = "black"),
  stat_cor(aes(nrel, nrel_mpif), method = "pearson", color = "black"),
  labs(x = "Fraction in scRNA",
       y = "Fraction in mpIF")
)

p1 <- ggplot(markers_pos_scrna_mpif_state_cd45n) +
  geom_point(aes(nrel, nrel_mpif, color = tumor_supersite)) +
  scale_color_manual(values = clrs$tumor_supersite) +
  common_layers

p2 <- ggplot(markers_pos_scrna_mpif_state_cd45n) +
  geom_point(aes(nrel, nrel_mpif, color = consensus_signature)) +
  scale_color_manual(values = clrs$consensus_signature) +
  common_layers

plot_grid(p1, p2, ncol = 1)

```

```{r chunk_220}

mpif_cell_state_n %>% 
  group_by(cell_type, fov_id) %>% 
  mutate(n_ct = sum(n),
         nrel_ct = sum(nrel)) %>% 
  ungroup() %>% 
  distinct(-n_ct, -nrel_ct, .keep_all = T) %>% 
  # sample_frac(0.01) %>% 
  ggplot(aes(fov_id, nrel_ct)) +
  geom_point(aes(patient_id, nrel_ct, color = tumor_supersite), 
             size = 0.01) + 
  facet_grid(~cell_type, 
             scale = "free_x", space = "free_x") +
  remove_xaxis +
  scale_color_manual(values = clrs$tumor_supersite)

```

